---
title: A gene signature consisting of ubiquitin ligases and deubiquitinating enzymes of SKP2 is associated with clinical outcome in breast cancer
output:
  word_document: default
  pdf_document:
    fig_caption: yes
  html_document:
    fig_caption: yes
---

# Setup
```{r eval=TRUE, include=FALSE}
# Load required libraries.
library(limma)
library(cgdsr)
library(ggplot2)
library(reshape2)
library(TCGA2STAT)
library(ggpubr)
library(ggsci)
library(plyr)
library(dplyr)
library(survminer)
library(egg)
library(gridExtra)
library(plotly)
library(stringr)
library(MASS)
library(scales)
library(flextable)
library(officer)
library(tibble)
library(jetset)
library(hgu133plus2.db)
library(DrInsight)
library(tidyr)
library(eulerr)
library(xtable)
library(readxl)
source("TumorNormalMatch2.R")

# Set the cancers and genes.
source("cancers_and_genes.R")
cancer = "BRCA"
# Set up a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

# For converting from HUGO names to common names later.
dict = list(CDKN1A="p21", CDKN1B="p27")
downstream_gene = "CDKN1B"

dub_genes = c("USP10", "USP13", "USP14") # USP7: control.
single_genes = c("SKP2") # c(ub_genes, dub_genes, "SKP2")

# Graphical parameters.
# https://stackoverflow.com/questions/39508304/margins-between-plots-in-grid-arrange
source("figure_graphical_parameters.R")
margin = theme(plot.margin = unit(c(2,2,2,2), "cm"))
```

# Results
## Section 1: Identification and validation of a gene signature to approximate SKP2 ubiquitination and protein levels
### Table 1
```{r echo = FALSE, eval = TRUE, fig.cap=""}
# Set up the columns.
sig_group = c("High", "Intermediate", "Low")
cn_ratios = c("FZR1 > USP10/USP13", "FZR1 = USP10/USP13", "FZR1 < USP10/USP13")
expected_SKP2_prot_lvls = c("Low", "Intermediate", "High")
expected_p27_prot_lvls = c("High", "Intermediate", "Low")

# Put into a data frame. 
table_1 = as.data.frame(cbind(sig_group, cn_ratios, expected_SKP2_prot_lvls, expected_p27_prot_lvls))
colnames(table_1) = c("SKP2 ubiquitination signature group", "FZR1 copy number vs USP10/USP13 copy number",
"Expected SKP2 protein levels", "Expected p27 protein levels")

# Turn into a flextable. 
table_1_ft = flextable(table_1)

autofit(table_1_ft)
```

### Figure 2
```{r echo = FALSE, eval = TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

# Divide the patients into the four molecular subtypes as listed here: https://ww5.komen.org/BreastCancer/SubtypesofBreastCancer.html. 
# On using .$ to extract column as vector: https://stackoverflow.com/a/27160317 
clin_dat = read.delim("brca_tcga_clinical_data.tsv", stringsAsFactors = F)
subtypes = list()
subtypes[["All"]] = clin_dat %>% .$Sample.ID
subtypes[["Luminal"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive") %>% .$Sample.ID
subtypes[["HER2"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Positive") %>% .$Sample.ID
#subtypes[["Non-TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive" | PR.status.by.ihc=="Positive" | IHC.HER2=="Positive") %>% .$Sample.ID
subtypes[["TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Negative") %>% .$Sample.ID

# CNA
# For each subtype ... 
plots = list()
dat_tables = list()
prot_name = dict[[downstream_gene]]
for(subtype in names(subtypes)) {
  # Set up the data frame that will hold the data for all the cancers, for subtype "subtype". 
  dat = data.frame(Genes=character(), ProteinLevels=numeric(), Group=character())
  # Get the sample IDs for this subtype.
  sample_IDs = base::gsub("\\-", "\\.", subtypes[[subtype]])
  
  # For each gene in dub_genes, 1) get for all the samples in the subtype a) the copy numbers of the dub gene, b) the copy numbers of the ub gene, and c) the protein levels of p27; 2) split the samples into two groups by copy-number status (copy number of dub gene > copy number of ub gene and v.v.); 3) compare the p27 levels between the two groups. 
  for(dub_gene in dub_genes) {
    # Get all the samples that have target protein expression.
    #prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_rppa"))[,1,drop=FALSE]
    prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_all"))[,1,drop=FALSE]
    # Remove all samples that do not have expression data.
    prot = prot[complete.cases(prot),,drop=FALSE]
    # if(nrow(prot)==0) next # If no protein data, skip to the next gene.
    
    # Get the CNA data of the predictor genes.
    cna = getProfileData(mycgds, c(dub_gene, ub_genes[1]), paste0(tolower(cancer), "_tcga_gistic"), paste0(tolower(cancer), "_tcga_cna"))
    # Filter out incomplete cases. 
    # Remove incomplete cases (NAs or NaNs) and log-transform.
    cna = cna[complete.cases(cna),,drop=FALSE]
    if(nrow(cna)==0) next # If no cna data, skip to the next gene. 
    
    # Keep only the samples that have both gene and protein levels. 
    good_samples = intersect(intersect(rownames(cna), rownames(prot)), sample_IDs)
    prot = prot[which(rownames(prot) %in% good_samples),,drop=FALSE]
    cna = cna[which(rownames(cna) %in% good_samples),,drop=FALSE]
    # if(nrow(prot)==0 | nrow(cna)==0) next 
    
    # Make sure the samples are in the same order in both the protein and gene copy number vectors.
    prot = prot[order(rownames(prot)),,drop=FALSE]
    cna = cna[order(rownames(cna)),,drop=FALSE]
    # identical(rownames(prot), rownames(cna))
    # Make sure the column names of the predictor genes are in alphabetical order.
    cna = cna[,order(colnames(cna), decreasing = FALSE),drop=FALSE]
    
    # Divide samples by copy number of the predictor genes.
    # Create the signature matrix.
    group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
    for(k in 1:nrow(cna)) {
      if(cna[k,1] < cna[k,2]) { group[k] = paste0("Low ubiquitination")} 
      else if(cna[k,1] > cna[k,2]) { group[k] = paste0("High ubiquitination")}
      else { group[k] = paste0("Intermediate ubiquitination")}
      
    }
    # Bind the grouping to the matrix.
    cna$Group = as.factor(group)
    # Keep only complete cases.
    cna = cna[complete.cases(cna),]
    prot = prot[rownames(prot) %in% rownames(cna),,drop=FALSE]
    if(identical(rownames(cna), rownames(prot))) {
      prot$Group = cna$Group
    }
    
    dat = rbind(dat, data.frame(Genes=paste0(dub_gene, ", ", ub_genes[1]), ProteinLevels=prot[,1], Group=prot$Group))
  }
   # Save the counts.
  counts_table_final = rbind(counts_table_final, data.frame(
    Gene = gene,
    Subtype = subtype,
    Count = nrow(dat),
    stringsAsFactors = F
    ))
  
  # Save dat to the list.
  dat_tables[[subtype]] = dat
  write.table(dat, file = paste0("Figure2_CNA_", subtype, ".csv"), sep = ",", row.names = F, col.names = F)
  
  # Use ggplot2 to create a dotplot. 
   plots[[subtype]] = ggplot(data=dat, aes(x=Genes, y=ProteinLevels, fill=Group, group=interaction(Group, Genes))) + # Each separate boxplot will be the "intersection" (i.e. interaction) of a gene (or gene signature) x group.
    geom_boxplot(lwd = boxplot_lwd) + # geom_violin() # https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size)) +
    scale_fill_manual(values=c("#0073C2FF", "#868686FF", "#EFC000FF")) + #scale_fill_jco() +
    labs(title=subtype, x=" ", y=" ") #+ 
    #stat_compare_means(label =  "p.signif", 
    #                   label.x = 1.5, 
    #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
    #                                      symbols = c("****", "***", "**", "*", " ")), 
    #                   size = p_value_size )  
}

# Arrange plots.
figure = ggpubr::ggarrange(plotlist = plots, 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
# Annotate figure by adding a common label.
# https://github.com/kassambara/ggpubr/issues/78
figure_2 = annotate_figure(figure,
                                   top = text_grob(paste(prot_name, "protein levels between SKP2 ubiquitination signature groups"), size = overall_title_size),
                                   left = text_grob(paste(prot_name, "protein levels"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Gene(s)"), size = common_axis_text_x_size),
                                   fig.lab = " ", 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
)

# Save to EPS.
setEPS()
postscript("Figure_2.eps", width = 25, height = 20)
figure_2
dev.off()

# Statistics.
# http://www.sthda.com/english/wiki/one-way-anova-test-in-r
stats = list()
for(subtype in names(subtypes)) {
  dat = dat_tables[[subtype]]
  for(gene_set in unique(dat$Genes)) {
    dat_sub = dat %>% dplyr::filter(Genes==gene_set) %>% droplevels()
    
    # Compute summary statistics.
    summary_stats = dat_sub %>% group_by(Group) %>%
      summarise(
        Count = n(),
        Mean = mean(ProteinLevels, na.rm = TRUE) %>% signif(digits = 2),
        SD = sd(ProteinLevels, na.rm = TRUE) %>% signif(digits = 2)
        )
    # Save to stats list.
    stats[[subtype]][[gene_set]][["Summary"]] = summary_stats
    
    # Perform ANOVA.
    res_aov = aov(ProteinLevels ~ Group, data = dat_sub)
    # Summary of the analysis.
    aov_stats = summary(res_aov)
    # Save to stats list.
    stats[[subtype]][[gene_set]][["Inferential"]] = aov_stats
    
  }
}
saveRDS(stats, file = "Figure2_statistics.rds")

# show_col(pal_jco()(4))
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
plotlist = list(figure_2)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Figure 3
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

# Divide the patients into the four molecular subtypes as listed here: https://ww5.komen.org/BreastCancer/SubtypesofBreastCancer.html. 
# On using .$ to extract column as vector: https://stackoverflow.com/a/27160317 
clin_dat = read.delim("brca_tcga_clinical_data.tsv", stringsAsFactors = F)
subtypes = list()
subtypes[["All"]] = clin_dat %>% .$Sample.ID
subtypes[["Luminal"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive") %>% .$Sample.ID
subtypes[["HER2"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Positive") %>% .$Sample.ID
#subtypes[["Non-TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive" | PR.status.by.ihc=="Positive" | IHC.HER2=="Positive") %>% .$Sample.ID
subtypes[["TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Negative") %>% .$Sample.ID

# Split patients into high and low ubiquitination signatures.
# Get the CNA data of the predictor genes.
cancer = "BRCA"
cna = getProfileData(mycgds, c("USP10", "FZR1"), paste0(tolower(cancer), "_tcga_gistic"), paste0(tolower(cancer), "_tcga_cna"))
# Filter out incomplete cases. 
cna = cna[complete.cases(cna),,drop=FALSE]

# Create the signature matrix.
group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
for(k in 1:nrow(cna)) {
  if(cna[k,1] < cna[k,2]) { group[k] = paste0("Low ubiquitination")} 
  else if(cna[k,1] > cna[k,2]) { group[k] = paste0("High ubiquitination")}
  else { group[k] = NA}
}
# Bind the grouping to the matrix.
cna$Group = as.factor(group)
# Keep only complete cases.
cna = cna[complete.cases(cna),]
rownames(cna) = as.character(base::gsub("\\.", "-", rownames(cna)))

# Get the expression data. 
if(file.exists("brca_tcga_rnaseq2.rds")) {
  exprs = readRDS("brca_tcga_rnaseq2.rds")
} else {
  exprs = getTCGA(disease="BRCA", data.type="RNASeq2")$dat
  # Save the expression data for later use. 
  saveRDS(exprs, "brca_tcga_rnaseq2.rds")
}
# Log-transform.
exprs = apply(exprs, MARGIN = 1, function(x) log2(x+1))
exprs = t(exprs) # The heck?? Why did apply() flip it??
# Keep only the primary tumors. 
type = sapply(colnames(exprs), function(s) unlist(strsplit(s, "-"))[4]) # Split the sample name up by dashes. The sample type is the fourth number (e.g. "01A") in the sample name.
exprs =  exprs[, base::grep("^01", type)] # The primary tumors are the ones whose type has "01" in it. 
# Shorten the sample names so we can match them with those in the cna data frame. 
colnames(exprs) = substr(colnames(exprs), 1, 15)
# Make sure it matches the cna matrix.
common_samples = intersect(colnames(exprs), rownames(cna))
exprs = exprs[,colnames(exprs) %in% common_samples]
cna = cna[rownames(cna) %in% common_samples,]
exprs = exprs[,order(colnames(exprs))]
cna = cna[order(rownames(cna)),]
identical(colnames(exprs), rownames(cna))

# Get the mRNA levels of SKP2.
mrna = getProfileData(mycgds, c("SKP2"), paste0(tolower(cancer), "_tcga_rna_seq_v2_mrna"), paste0(tolower(cancer), "_tcga_rna_seq_v2_mrna"))
# Filter out incomplete cases. 
mrna = mrna[complete.cases(mrna),,drop=FALSE]
rownames(mrna) = as.character(base::gsub("\\.", "-", rownames(mrna)))

plots = list()
cor_table = data.frame(
  Subtype = character(),
  NumSamples = integer(),
  NumGenes = integer(),
  CorrelationCoefficient = numeric(),
  Significance = numeric()
)
for(subtype in names(subtypes)) {
  # Get the sample IDs for this subtype.
  sample_IDs = subtypes[[subtype]]
  # Subset exprs and cna.
  exprs_subtype = exprs[,colnames(exprs) %in% sample_IDs,drop=F]
  cna_subtype = cna[rownames(cna) %in% sample_IDs,,drop=F]
  # The samples should still be the same in both exprs_subtype and cna_subtype since these are subsets of exprs and cna, and exprs and cna were already shown to have the same samples--see the line identical(colnames(exprs), rownames(cna)).
  
  # Perform differential expression analysis using limma.
  # Create the design matrix. 
  # Set the high-ubiquitination signature as the baseline (=1.) For the low-vs-high-ubiquitination column, 0 = high ubiquitination (i.e. no change from the baseline) and 1 = low ubiquitination (i.e. a change from the baseline.)
  design_ub = cbind(HighUb=1, LowVsHighUb=ifelse(cna_subtype$Group=="High ubiquitination", 0, 1))

  # Fit the model.
  fit_ub = lmFit(exprs_subtype, design_ub)
  fit_ub = eBayes(fit_ub)
  topTable(fit_ub)
  limmares_ub = topTable(fit_ub, coef = 2, n = Inf, sort ="p")

  # Split patients into high and low SKP2 groups.
  # Subset mrna.
  mrna_subtype = mrna[rownames(mrna) %in% sample_IDs,,drop=F]
  # Create the signature matrix.
  group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
  high_cutoff = 0.5
  low_cutoff = 0.5
  for(k in 1:nrow(mrna_subtype)) {
    if(mrna_subtype[k,1] >= quantile(mrna_subtype[,1], high_cutoff, na.rm=TRUE)) { group[k] = paste0("High SKP2")} 
    else if(mrna_subtype[k,1] < quantile(mrna_subtype[,1], low_cutoff, na.rm=TRUE)) { group[k] = paste0("Low SKP2")}
    else { group[k] = NA}
  }
  # Bind the grouping to the matrix.
  mrna_subtype$Group = as.factor(group)

  # Make sure it matches the mrna_subtype matrix.
  common_samples = intersect(colnames(exprs), rownames(mrna_subtype))
  exprs_subtype = exprs[,colnames(exprs) %in% common_samples]
  mrna_subtype = mrna_subtype[rownames(mrna_subtype) %in% common_samples,]
  exprs_subtype = exprs_subtype[,order(colnames(exprs_subtype))]
  mrna_subtype = mrna_subtype[order(rownames(mrna_subtype)),]
  identical(colnames(exprs_subtype), rownames(mrna_subtype))

  # Perform differential expression analysis using limma.
  # Create the design matrix. 
  # Set the high-ubiquitination signature as the baseline (=1.) For the low-vs-high-ubiquitination column, 0 = high ubiquitination (i.e. no change from the baseline) and 1 = low ubiquitination (i.e. a change from the baseline.)
  design_SKP2_mrna = cbind(LowSKP2=1, HighVsLowSKP2Ub=ifelse(mrna_subtype$Group=="Low SKP2", 0, 1))

  # Fit the model.
  fit_SKP2_mrna = lmFit(exprs_subtype, design_SKP2_mrna)
  fit_SKP2_mrna = eBayes(fit_SKP2_mrna)
  topTable(fit_SKP2_mrna)
  limmares_SKP2_mrna = topTable(fit_SKP2_mrna, coef = 2, n = Inf, sort ="p")

  # Order the limma results tables by rowname (gene names.)
  limmares_SKP2_mrna = limmares_SKP2_mrna[order(rownames(limmares_SKP2_mrna)),]
  limmares_ub = limmares_ub[order(rownames(limmares_ub)),]
  identical(rownames(limmares_SKP2_mrna), rownames(limmares_ub))
  
  # Get the t-statistics. 
  #t_SKP2_mrna = limmares_SKP2_mrna$t
  #t_ub = limmares_ub$t
  # Get the log fold change.
  lfc_SKP2_mrna = limmares_SKP2_mrna$logFC
  lfc_ub = limmares_ub$logFC
  
  # Calculate the Spearman correlation coefficient. 
  cor_test = cor.test(lfc_SKP2_mrna, lfc_ub, method = "spearman")
  rho = signif(cor_test$estimate, 2)
  p_val = ifelse(cor_test$p.value == 0, 2.2e-16, formatC(cor_test$p.value, format = "e", digits = 2)) #signif(cor_test$p.value, 2)
  
  # Add to table. 
  cor_table = rbind(cor_table, data.frame(
  Subtype = subtype,
  NumSamples = nrow(cna_subtype),
  NumGenes = length(lfc_SKP2_mrna),
  CorrelationCoefficient = rho,
  Significance = p_val
))
  rm(cna_subtype)
  gc()
  
  # Make a combined topTable and save to file.
  colnames(limmares_ub) = paste0("UbiquitinationSignature_", colnames(limmares_ub))
  colnames(limmares_SKP2_mrna) = paste0("SKP2Expression_", colnames(limmares_SKP2_mrna))
  topTable_combined = cbind(limmares_ub, limmares_SKP2_mrna) %>% as.data.frame %>% .[order(rownames(.)),]
  write.table(topTable_combined, paste0("topTable_", subtype, ".csv"), sep = ",")
  
  # Graph. 
  d = data.frame(SKP2=lfc_SKP2_mrna, Ub=lfc_ub)
  rho_y = max(lfc_ub)
  #if(cor_test$p.value > 0) {
  #  nonmetric_label = c(paste("ρ==", rho)
  #                      #, paste("p==", p_val)
  #                      )
  #} else {
  #  nonmetric_label =  c(paste("ρ==", rho)
  #                       #,paste("p<", p_val)
  #                       )
  #}
  figure = ggplot(d, aes(x = SKP2, y = Ub)) +
    geom_point(color = "#868686FF") +
    geom_smooth(method = lm, se = FALSE, color = "#CD534CFF") +
    theme_classic() +
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
    labs(title=subtype, x=" ", y=" ") + 
    annotate("text", 
             x = 0, 
             y = rho_y, # c(rho_y, 0.88*rho_y), 
             size = p_value_size,
             label = paste("ρ==", rho),#nonmetric_label, #as.expression(bquote("Chi-square"~italic(p)==bold(.(pval))))
             parse = TRUE)
  
    plots[[subtype]] = figure
  
    # Save the counts.
    #counts_table_final = rbind(counts_table_final, data.frame(
    #  Gene = gene,
    #  Subtype = subtype,
    #  Count = nrow(dat),
    #  stringsAsFactors = F
    #  ))
    #counts_table_final = rbind(counts_table_final, data.frame(
    #  Gene = gene,
    #  Subtype = subtype,
    #  Count = nrow(dat),
    #  stringsAsFactors = F
    #  ))
}

#saveRDS(cor_table, file = "Figure3_statistics.rds")

# Arrange plots.
figure_3 = ggpubr::ggarrange(plotlist = plots, 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20)
                   )
# Annotate figure by adding a common label.
figure_3 = annotate_figure(figure_3,
                              fig.lab = " ", 
                              fig.lab.size = figure_label_size,
                              fig.lab.face = figure_label_face,
                              top = text_grob(paste("Correlation between log fold changes (LFCs) of genes \n between ubiquitination signature groups and between SKP2 mRNA groups"), size = overall_title_size),
                              left = text_grob(paste("LFCs of genes between ubiquitination signature groups"), rot = 90, size = common_axis_text_y_size),
                              bottom = text_grob(paste("LFCs of genes between SKP2 mRNA groups"), size = common_axis_text_x_size)
  )

# Save to EPS.
#setEPS()
#postscript("Figure_3.eps", width = 20, height = 20)
#figure_3
#dev.off()

# Save the rownames.
#write.table(rownames(exprs), "BRCA_TCGA_RNASeq2_gene_names.tsv", row.names = F, col.names = F, sep = "\t", quote = F)
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
plotlist = list(figure_3)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

## Section 2: Assessment of the clinical relevance of our SKP2 ubiquitination signature
### Figure 4
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("USP10", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")
# Get the expression data.
cna_subset = getProfileData(mycgds, genes_sig, "brca_metabric_cna", "brca_metabric_cna")
# Get the clinical data.
clin_sig = getClinicalData(mycgds, "brca_metabric_cna")

# Throw out all the samples that don't have expression data.
cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
# Subset to include only the patients with no chemo, hormone, or radiation therapy. 
# clin_sig = clin_sig[clin_sig$CHEMOTHERAPY=="NO" & clin_sig$HORMONE_THERAPY=="NO" & clin_sig$RADIO_THERAPY=="NO",,drop=FALSE]
# Make sure both the clinical and expression data have the same patients_
overlap = intersect(rownames(clin_sig), rownames(cna_subset))
clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
# Order rownames so they're both in the same order.
cna_subset = cna_subset[order(rownames(cna_subset)),]
clin_sig = clin_sig[order(rownames(clin_sig)),]
identical(rownames(cna_subset), rownames(clin_sig))

clin_sig$SAMPLE_ID = rownames(clin_sig)
subtypes = list()
subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
#subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID

plots = list()
stats = list()
for(subtype in names(subtypes)) {
  # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
  sample_IDs = subtypes[[subtype]]
  cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,]
  clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
  
  ## Divide up the samples by copy number of the genes in the signature. 
  cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample_
  # for(k in 1:nrow(cna_subset)) {
  #      if(cna_subset[k,1] < cna_subset[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
  #      else if(cna_subset[k,1] > cna_subset[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")}
  #      else { cna_subset_group[k] = NA}
  #      
  #    }
  for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
      else if(cna_subset_subtype[k,1] > cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")} 
      else { cna_subset_group[k] = paste0("Intermediate ubiquitination") }
      
    }
  cna_subset_subtype$Group = cna_subset_group
  # Remove NAs.
  cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
  clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
  identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))

  # Re-order the levels of "Group": Low - Intermediate - High.
  cna_subset_subtype$Group = factor(cna_subset_subtype$Group, levels = c('High ubiquitination', 'Intermediate ubiquitination', 'Low ubiquitination'))
  
  ## Perform survival analysis.
  # Create the survival matrix.
  surv_matrix_sig = clin_sig_subtype[,which(colnames(clin_sig_subtype) %in% c("OS_MONTHS", "VITAL_STATUS"))]
  surv_matrix_sig$VITAL_STATUS = as.numeric(ifelse(surv_matrix_sig$VITAL_STATUS=="Died of Disease", 1, 0))
  identical(rownames(surv_matrix_sig), rownames(cna_subset_subtype))
  surv_matrix_sig$Group = cna_subset_subtype$Group # Make sure the clinical and expression samples are IDENTICAL (same samples in the same order) before doing this! 
  # Create the survival object.
  surv_obj_sig = survival::Surv(time = surv_matrix_sig$OS_MONTHS, event = surv_matrix_sig$VITAL_STATUS)
  # Fit the survival curves.
  surv_model_sig = survival::survfit(surv_obj_sig ~ Group, data = surv_matrix_sig)

  # Save to list.
  stats[[subtype]] = surv_model_sig
  
  # Plot.
  ggsurv_cna_ub = ggsurvplot(surv_model_sig, 
           data = surv_matrix_sig, 
           font.tickslab = tick_label_size,
           size = km_line_size, # Line size.
           pval = TRUE,
           pval.size = p_value_size,
           conf.int = F, # Show confidence intervals for point estimates of survival curves.
           break.time.by = 50,
           title = subtype,
           xlab = " ",   # Customize X axis label.
           ylab = " ",
           legend.title = "Group",
           legend.labs = c('High ubiquitination', 'Intermediate ubiquitination', 'Low ubiquitination'),
           risk.table = FALSE, # Set to TRUE to turn risk tables on. 
           risk.table.y.text.col = T, # Colour risk table text annotations.
           risk.table.y.text = FALSE, # Show bars instead of names in text annotations in legend of risk table)
           palette = c("#0073C2FF", "#868686FF", "#EFC000FF")
           ) 

  # https://github.com/kassambara/survminer/issues/2
 ggsurv_cna_ub$plot = ggsurv_cna_ub$plot +
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size))
  #ggsurv_cna_ub$table = ggsurv_cna_ub$table + labs(title = "Number at risk") # Uncomment this line when turning risk tables back on. 
  plots[[subtype]] = ggsurv_cna_ub$plot
  # print(ggsurv_cna_ub)
 
  # Save the counts.
  counts_table_final = rbind(counts_table_final, data.frame(
    Gene = sig_gene,
    Subtype = subtype,
    Count = sum(surv_model_sig$n),
    stringsAsFactors = F
    )) 
}

# Save statistics to file. 
saveRDS(stats, file = "Figure4a_statistics.rds")

# Arrange plots. 
figure = ggpubr::ggarrange(plotlist = plots, 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
  
# Annotate arranged ggsurvplots with common labels. 
figure_4a = annotate_figure(figure,
                                   top = text_grob("Overall survival of breast-cancer patients by SKP2-ubiquitination signature", size = overall_title_size),
                                   left = text_grob(paste("Survival probability"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Survival time (months)"), size = common_axis_text_x_size),
                                   fig.lab = "a", 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
)


### GSE17705
signature_genes = c("FZR1", "USP10")
## Load the data.
eset = readRDS("GSE17705.rds")
exprs = exprs(eset)
pData = pData(eset)
rm(eset)
gc()

## Create the survival matrix.
# Columns to keep:
keep = c(41, 44)
surv.matrix = pData[,keep,drop=FALSE]
# Rename the columns.
colnames(surv.matrix) = c("DR_event", "DR_time")
## Create the matrix of genes (predictors).
# Use hgu133plus2.db to map Affy probe IDs to gene symbols.
mapping = AnnotationDbi::select(hgu133plus2.db, rownames(exprs), "SYMBOL")
mapping = mapping[which(mapping$SYMBOL %in% signature_genes),]

# Now let's pick the probesets using jetset.
mapping$SYMBOL = as.factor(mapping$SYMBOL)
predictors.list = list()
probesets.final = c()
for(i in levels(mapping$SYMBOL)) {
  # Check how many probesets there are for a given gene (symbol.)
  if(sum(mapping$SYMBOL==i)==1) {
    # If there is only one probeset, I guess we'll just have to use that one?
    # Get the probeset.
    probeset = mapping[which(mapping$SYMBOL==i),1]
    # Add to the list the expression of that probeset across all patients.
    predictors.list[[i]] = exprs[which(rownames(exprs)==probeset),]
    # Keep track of which probeset was used! 
    probesets.final = c(probesets.final, probeset)
    
  } else {
    # We will use jetset to choose.
    jscores = jscores('hgu133plus2', symbol=i)
    
    # Check that the best jetset probeset is actually in the expression matrix.
    best.probeset.in.exprs = FALSE
    nth.best.probeset = 1
    while(best.probeset.in.exprs==FALSE) {
      best.probeset = rownames(jscores[nth.best.probeset,])
      
      if(nrow(exprs[which(rownames(exprs)==best.probeset),,drop=FALSE]) > 0) {
        # Add to the list the expression of that probeset across all patients.
        predictors.list[[i]] = exprs[which(rownames(exprs)==best.probeset),]
        # Keep track of which probeset was used! 
        probesets.final = c(probesets.final, best.probeset)
        # Stop the while() loop.
        best.probeset.in.exprs = TRUE
        
      } else {
        # The best jetset probeset is not in the expression matrix. 
        # Get the next best probeset.
        nth.best.probeset = nth.best.probeset + 1
        
      } # End else: the best jetset probeset is not in the expression matrix.
      
    } # End while loop to choose the nth best jetset probeset. 
    
  } # End else: there are >1 probesets for gene i.
  
} # End for() loop.

# Make matrix of predictor variables (the above genes).
predictors = as.data.frame(t(do.call(rbind, predictors.list)))

## Divide samples by expression of the predictor genes.
# Create the signature matrix.
upper.quantile = 0.5
lower.quantile = 0.5
group = c() # The vector that holds the grouping (high vs. low expression) for each sample.
 for(k in 1:nrow(predictors)) {
  if(predictors[k,1] <= quantile(predictors[,1], lower.quantile, na.rm=TRUE) & predictors[k,2] > quantile(predictors[,2], upper.quantile, na.rm=TRUE)) { group[k] = "Low SKP2 ubiquitination"} 
   else if(predictors[k,1] > quantile(predictors[,1], upper.quantile, na.rm=TRUE) & predictors[k,2] <= quantile(predictors[,2], lower.quantile, na.rm = TRUE)) { group[k] = "High SKP2 ubiquitination"}
   else { group[k] = "Intermediate SKP2 ubiquitination"}
  
 }
#for(k in 1:nrow(predictors)) {
#  if(predictors[k,1] > quantile(predictors[,1], upper.quantile, na.rm=TRUE) & #predictors[k,2] <= quantile(predictors[,2], lower.quantile, na.rm = TRUE)) { group[k] = "High SKP2 ubiquitination"} 
#   else { group[k] = "Intermediate/low SKP2 ubiquitination"}
  
# }
predictors$Group = group
# Bind the grouping to the *survival* matrix, but only if it's in the same order as the predictors matrix.
# Keep only complete cases.
predictors = predictors[complete.cases(predictors),]
surv.matrix = surv.matrix[rownames(surv.matrix) %in% rownames(predictors),,drop=FALSE]
predictors = predictors[order(rownames(predictors)),]
surv.matrix = surv.matrix[order(rownames(surv.matrix)),]
if(identical(rownames(predictors), rownames(surv.matrix))) {
  surv.matrix$Group = predictors$Group
}

# Re-order the levels of "Group": High - Intermediate - Low.
surv.matrix$Group = factor(surv.matrix$Group, levels = c('High SKP2 ubiquitination', 'Intermediate SKP2 ubiquitination', 'Low SKP2 ubiquitination'))
  
## Perform survival analysis.
# Create the survival object.
surv.matrix$DR_time = as.numeric(surv.matrix$DR_time)
surv.matrix$DR_event = as.numeric(surv.matrix$DR_event)
surv.obj = survival::Surv(time = surv.matrix$DR_time, event = surv.matrix$DR_event)
# Fit the survival curves.
surv.model = survival::survfit(surv.obj ~ Group, data = surv.matrix)
counts = sum(surv.model$n)
  
# Save statistics to file. 
saveRDS(surv.model, file = "Figure4b_statistics.rds")

# Plot.
ggsurv = ggsurvplot(surv.model, 
           data = surv.matrix, 
           font.tickslab = tick_label_size,
           size = km_line_size, # Line size.
           pval = TRUE,
           pval.size = p_value_size,
           conf.int = F, # Show confidence intervals for point estimates of survival curves.
           #break.time.by = 50,
           title = " ",
           xlab = " ",   # Customize X axis label.
           ylab = " ",
           legend = "right",
           legend.title = "Group",
           legend.labs = c("High ubiquitination", "Intermediate ubiquitination", "Low ubiquitination"),
           risk.table = FALSE, # Set to TRUE to turn risk tables on. 
           risk.table.y.text.col = T, # Colour risk table text annotations.
           risk.table.y.text = FALSE, # Show bars instead of names in text annotations in legend of risk table)
           palette = c("#0073C2FF", "#868686FF", "#EFC000FF")
           )  

# https://github.com/kassambara/survminer/issues/2
ggsurv$plot = ggsurv$plot +
  theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size))

# Annotate arranged ggsurvplots with common labels. 
figure_4b = annotate_figure(ggsurv$plot,
                                   top = text_grob("Distant recurrence–free survival by ubiquitination group in ER-positive BC patients treated with tamoxifen", size = overall_title_size),
                                   left = text_grob(paste("Survival probability"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Survival time (years)"), size = common_axis_text_x_size),
                                   fig.lab = "b", 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
)

plots_fig4 = list()
plots_fig4$Fig4a = figure_4a 
plots_fig4$Fig4b = figure_4b
figure_4 = ggpubr::ggarrange(plotlist = plots_fig4,
                             nrow = 2,
                             ncol = 1,
                             hjust = c(-0.5, -1, -0.5, -1, -1.5))

# Save to EPS.
setEPS()
postscript("Figure_4.eps", width = 20, height = 20)
figure_4
dev.off()
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 25, fig.margin = TRUE, fig.cap = ""}
figure_4
```

### Figures 5 and 6
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  ClinicalVariable = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("USP10", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")
# Get the expression data.
cna_subset = getProfileData(mycgds, genes_sig, "brca_metabric_cna", "brca_metabric_cna")
# Get the clinical data.
clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
clin_sig$SAMPLE_ID = rownames(clin_sig)
# Throw out all the samples that don't have cna data.
cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
# Make sure both the clinical and expression data have the same patients_
overlap = intersect(rownames(clin_sig), rownames(cna_subset))
clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
# Order rownames so they're both in the same order.
cna_subset = cna_subset[order(rownames(cna_subset)),]
clin_sig = clin_sig[order(rownames(clin_sig)),]
identical(rownames(cna_subset), rownames(clin_sig))

clin_sig$SAMPLE_ID = rownames(clin_sig)
subtypes = list()
subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
#subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID

plots = list()
results = list()
for(subtype in names(subtypes)) {
  # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
  sample_IDs = subtypes[[subtype]]
  cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,]
  clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
  
  ## Divide up the samples by copy number of the genes in the signature. 
  cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample_
  # for(k in 1:nrow(cna_subset)) {
  #      if(cna_subset[k,1] < cna_subset[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
  #      else if(cna_subset[k,1] > cna_subset[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")}
  #      else { cna_subset_group[k] = NA}
  #      
  #    }
  for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
      else if(cna_subset_subtype[k,1] > cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")} 
      else { cna_subset_group[k] = paste0("Intermediate ubiquitination") }
      
    }
  cna_subset_subtype$Group = cna_subset_group
  # Remove NAs.
  cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
  clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
  identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))
  
  # Create a table holding, for each sample, 1) SKP2 ubiquitination signature, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
  other_clin_features = data.frame(UbSig=cna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
  
  # Stage
  # Use chi-squared test. 
  confusion_stage = table(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
  chisq_test_stage = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
  # Graph.
  proptable_stage = prop.table(confusion_stage,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
  melted_stage = melt(proptable_stage)
  colnames(melted_stage) = c("UbSignature", "Stage", "Proportion")
  melted_stage$Stage = as.factor(melted_stage$Stage)
  # Get the y-value for the p-value annotation. 
  pval_y = min(max(melted_stage$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
  pval = as.character(signif(chisq_test_stage$p.value, 2))
  # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
  # Re-order the levels of "UbSignature": Low - Intermediate - High.
  melted_stage$UbSignature = factor(melted_stage$UbSignature, levels = c('High ubiquitination', 'Intermediate ubiquitination', 'Low ubiquitination'))
  
  # Plot.
  stage_plot = ggplot(melted_stage, aes(x = UbSignature, y = Proportion, fill = Stage)) +
    geom_bar(stat = "identity", position=position_dodge()) + #color="black",
    #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
    annotate("text", 
             x = 2, 
             y = pval_y, 
             size = p_value_size,
             label = paste("p==", pval), #as.expression(bquote("Chi-square"~italic(p)==bold(.(pval))))
             parse = TRUE) + # Annotate with p-value. 
    # https://www.r-bloggers.com/math-notation-for-r-plot-titles-expression-and-bquote/
    # https://r.789695.n4.nabble.com/lattice-plotmath-how-to-get-a-variable-in-bold-face-td3569599.html
    scale_fill_manual(breaks = c('0', '1', '2', '3', '4'),
                        values=pal_npg()(5)) +
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = (axis_text_x_size - 4), hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size)) +
    #scale_size(range = c(0, 10)) +
    labs(x=" ", y=" ", title=paste0(subtype))
    
  print(paste0("The p-value of the chi-squared test on the distribution of tumor stage by ubiquitination signature in subtype ", subtype, " is ", chisq_test_stage$p.value))
  
  # Grade
  # Use chi-squared test.
  confusion_grade = table(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
  chisq_test_grade = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
  # Graph.
  proptable_grade = prop.table(confusion_grade,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
  melted_grade = melt(proptable_grade)
  colnames(melted_grade) = c("UbSignature", "Grade", "Proportion")
  melted_grade$Grade = as.factor(melted_grade$Grade)
  # Get the y-value for the p-value annotation. 
  pval_y = min(max(melted_grade$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
  pval = as.character(signif(chisq_test_grade$p.value, 2))
  # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
  # Re-order the levels of "UbSignature": Low - Intermediate - High.
  melted_grade$UbSignature = factor(melted_grade$UbSignature, levels = c('High ubiquitination', 'Intermediate ubiquitination', 'Low ubiquitination'))
  
  # Plot.
  grade_plot = ggplot(melted_grade, aes(x = UbSignature, y = Proportion, fill = Grade)) +
    geom_bar(stat = "identity", position=position_dodge()) + 
    #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
    annotate("text", 
             x = 2, 
             y = pval_y, 
             size = p_value_size,
             label =  paste("p==", pval), # as.expression(bquote("Chi-square"~italic(p)==bold(.(pval)))) 
             parse = TRUE) + # Annotate with p-value. 
    #scale_x_continuous(breaks=seq(1, 3, 1)) +
    scale_fill_manual(breaks = c('1', '2', '3'),
                        values=pal_npg()(3)) +
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = (axis_text_x_size - 4), hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size)) +
    #scale_size(range = c(0, 10)) + 
    labs(x=" ", y=" ", title=paste0(subtype))
  
  print(paste0("The p-value of the chi-squared test on the distribution of tumor grade by ubiquitination signature in subtype ", subtype, " is ", chisq_test_grade$p.value))
  
  # Number of lymph nodes.
  # https://data.princeton.edu/wws509/r/overdispersion
  # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
  # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
  # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
  mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
  p_vals = coef(summary(mnb))[,4]
  
  # Add plots to plots list.
  plots[["Stage"]][[subtype]] = stage_plot
  plots[["Grade"]][[subtype]] = grade_plot
  # Add statistical test results to list. 
  results[["Stage"]][["Confusion"]][[subtype]] = confusion_stage
  results[["Stage"]][["Test"]][[subtype]] = chisq_test_stage
  results[["Grade"]][["Confusion"]][[subtype]] = confusion_grade
  results[["Grade"]][["Test"]][[subtype]] = chisq_test_grade
  results[["Nodes"]][["Test"]][[subtype]] = summary(mnb)
  
  # Save the counts.
  counts_table_final = rbind(counts_table_final, data.frame(
    Gene = sig_gene,
    Subtype = subtype,
    ClinicalVariable = "Stage",
    Count = sum(confusion_stage),
    stringsAsFactors = F
    ))
  counts_table_final = rbind(counts_table_final, data.frame(
    Gene = sig_gene,
    Subtype = subtype,
    ClinicalVariable = "Grade",
    Count = sum(confusion_grade),
    stringsAsFactors = F
    ))
}
saveRDS(results, file = "Figure5_6_statistics.rds")
saveRDS(plots, file = "Figure5_6_plots.rds")

i = 2
annotated_fig_list = list()
for(variable in c("Stage", "Grade")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[variable]], 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                   top = text_grob(paste("Association of SKP2 ubiquitination signature with", str_to_lower(full_variable)), size = overall_title_size),
                                   left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Ubiquitination signature group"), size = common_axis_text_x_size),
                                   fig.lab = " ", #paste0(str_to_lower(letters[i])), 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
                        )
    annotated_fig_list[[variable]] = figure_annotated
    
    i = i + 1
}
# Save to EPS.
setEPS()
postscript("Figure_5.eps", width = 20, height = 20)
annotated_fig_list$Stage
dev.off()

setEPS()
postscript("Figure_6.eps", width = 20, height = 20)
annotated_fig_list$Grade
dev.off()
```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Stage)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Grade)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Figure 7
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("USP10", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")
# Get the expression data.
cna_subset = getProfileData(mycgds, genes_sig, "brca_metabric_cna", "brca_metabric_cna")
# Get the clinical data.
clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
clin_sig$SAMPLE_ID = rownames(clin_sig)
subtypes = list()
subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
#subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
# Throw out all the samples that don't have cna data.
cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
# Make sure both the clinical and expression data have the same patients_
overlap = intersect(rownames(clin_sig), rownames(cna_subset))
clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
# Order rownames so they're both in the same order.
cna_subset = cna_subset[order(rownames(cna_subset)),]
clin_sig = clin_sig[order(rownames(clin_sig)),]
identical(rownames(cna_subset), rownames(clin_sig))

plots = list()
stats = list() # Holds the mean and variance of the counts for each subtype.
results = list() # Holds the negative binomial regression model of the counts for each subtype. 
for(subtype in names(subtypes)) {
  # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
  sample_IDs = subtypes[[subtype]]
  cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,]
  clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
  
  ## Divide up the samples by copy number of the genes in the signature. 
  cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample_
  # for(k in 1:nrow(cna_subset)) {
  #      if(cna_subset[k,1] < cna_subset[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
  #      else if(cna_subset[k,1] > cna_subset[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")}
  #      else { cna_subset_group[k] = NA}
  #      
  #    }
  for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
      else if(cna_subset_subtype[k,1] > cna_subset_subtype[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")} 
      else { cna_subset_group[k] = paste0("Intermediate ubiquitination") }
      
    }
  cna_subset_subtype$Group = cna_subset_group
  # Remove NAs.
  cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
  clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
  identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))
  
  # Create a table holding, for each sample, 1) SKP2 ubiquitination signature, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
  other_clin_features = data.frame(UbSig=cna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
  
  # Re-order the levels of "UbSig": High - Intermediate - Low.
  other_clin_features$UbSig = factor(other_clin_features$UbSig, levels = c('High ubiquitination', 'Intermediate ubiquitination', 'Low ubiquitination'))
  
  # Get the mean and variance of the counts.
  stats_table = data.frame(
    Subtype = subtype,
    n = nrow(cna_subset_subtype),
    "Mean number of positive lymph nodes" = round(mean(other_clin_features$NumLymphNodes, na.rm = T), 2),
    "Variance" = round(var(other_clin_features$NumLymphNodes, na.rm = T), 2)
  )
  colnames(stats_table) = base::gsub("\\.", " ", colnames(stats_table))
  stats[[subtype]] = stats_table
  
  # Build a negative binomial regression model.
  # https://data.princeton.edu/wws509/r/overdispersion
  # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
  # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
  # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
  mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
  p_vals = coef(summary(mnb))[,4]
  
  # Graph.
  # http://www.cookbook-r.com/Graphs/Plotting_distributions_(ggplot2)/
  lymph_nodes = other_clin_features[,c("UbSig", "NumLymphNodes")]
  nodes_plot = ggplot(lymph_nodes, aes(x=NumLymphNodes, fill=UbSig)) + 
    geom_histogram(position="identity", stat = "count") + 
    facet_grid(UbSig ~ ., margins = TRUE, scales = "free") + 
    #geom_boxplot() + 
    labs(fill = "Group") +
    #annotate("text", x = 4, y = 25, label = "italic(R) ^ 2 == 0.75", parse = TRUE) + # Annotate with p-value.
    scale_fill_manual(values=c("#0073C2FF", "#868686FF", "#EFC000FF", "#CD534CFF")) +
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size),
          strip.text.y = element_blank()) +
    labs(x=" ", y=" ", title=paste0(subtype))
  
  # Use chi-squared test? Or t-test? 
  #confusion_nodes = matrix(, nrow = length(levels(other_clin_features$UbSig)), ncol = 4)
  #i = 1
  #for(ubsig in levels(other_clin_features$UbSig)) {
  #  # Get the total number of lymph nodes for that ubiquitination signature.
  #  nodes_total = sum(other_clin_features[other_clin_features$UbSig==ubsig,"NumLymphNodes"], na.rm = T)
  #  avg_nodes = nodes_total / nrow(other_clin_features[other_clin_features$UbSig==ubsig,])
  #  confusion_nodes[i,] = c(ubsig, nodes_total, nrow(other_clin_features[other_clin_features$UbSig==ubsig,]), avg_nodes)
  #  i = i+1
  #}
  #colnames(confusion_nodes) = c("UbiquitinationSignature", "TotalNumPosNodes", "UbSigCount", "AvgNumPosNodes")
  ## Perform Kruskal-Wallis test.
  #kw_test = kruskal.test(NumLymphNodes ~ UbSig, data = other_clin_features)
  
  #print(paste0("The p-value of the Kruskal-Wallis test on the average number of positive lymph nodes per patient by ubiquitination signature in subtype ", subtype, " is ", kw_test$p.value))
  
  # Add plots to plots list.
  plots[["Nodes"]][[subtype]] = nodes_plot
  # Add statistical test results to list. 
  results[["Nodes"]][["Test"]][[subtype]] = summary(mnb)
  
  # Save the counts.
  counts_table_final = rbind(counts_table_final, data.frame(
    Gene = sig_gene,
    Subtype = subtype,
    Count = sum(!is.na(other_clin_features$NumLymphNodes)),
    stringsAsFactors = F
    ))
}
saveRDS(plots, file = "Figure7_plot.rds")
saveRDS(stats, file = "Figure7_descriptive_statistics.rds")
saveRDS(results, file = "Figure7_inferential_statistics.rds")

i = 1
annotated_fig_list = list()
for(variable in c("Nodes")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[variable]], 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                   top = text_grob(paste("Association of SKP2 ubiquitination signature with", str_to_lower(full_variable)), size = overall_title_size),
                                   left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Number of positive lymph nodes"), size = common_axis_text_x_size),
                                   fig.lab = paste0(" "), 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
                        )
    annotated_fig_list[[variable]] = figure_annotated
    
    i = i + 1
}
# Save to EPS.
setEPS()
postscript("Figure_7.eps", width = 20, height = 20)
annotated_fig_list$Nodes
dev.off()
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = " "}
plotlist = list(annotated_fig_list$Nodes)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Table 2
```{r echo = FALSE, eval=TRUE, include=FALSE}
# Build the table of results from the negative binomial model (for nodes.)
col_names = c("Factor level", "Estimate", "Std. error", "z-value", "p-value")
ft_list = list()
for(subtype in names(results$Nodes$Test)) {
  coef_formatted = as.data.frame(apply(coef(results$Nodes$Test[[subtype]]), c(1,2), signif, digits = 2))
 rownames(coef_formatted) = base::gsub("UbSig", "", rownames(coef_formatted))
 coef_formatted = add_column(coef_formatted, d = rownames(coef_formatted), .before = "Estimate")
 colnames(coef_formatted) = col_names
 
 significance = c()
 for(p_value in coef_formatted$`p-value`) {
   if(p_value < 0.001) {
     significance_i = "***"
   } else if(p_value >= 0.001 & p_value < 0.01) {
     significance_i = "**"
   } else if(p_value >= 0.01 & p_value < 0.05) {
     significance_i = "*"
   } else {
     significance_i = ""
   }
   
   significance = c(significance, significance_i)
 }
 coef_formatted$Significance = significance  
 ft = flextable(coef_formatted)
 ft = add_header_lines(ft, values = subtype, top = TRUE)
 ft_list[[subtype]] = autofit(ft)

}
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$All
ft_list$Luminal
ft_list$HER2
ft_list$TNBC
```

## Section 3: Use of the SKP2 ubiquitination signature to discover potential treatments for ER+ breast cancer
### Table 3
```{r echo=FALSE, eval=FALSE, include=FALSE}
if(!file.exists("Table_4_neg.rds")) {
  # Split patients into high and low ubiquitination signatures.
  # Get the CNA data of the predictor genes.
  cna = getProfileData(mycgds, c("USP10", "FZR1"), paste0(tolower(cancer), "_tcga_gistic"), paste0(tolower(cancer), "_tcga_cna"))
  # Filter out incomplete cases. 
  cna = cna[complete.cases(cna),,drop=FALSE]

  # Create the signature matrix.
  group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
  for(k in 1:nrow(cna)) {
    if(cna[k,1] < cna[k,2]) { group[k] = paste0("Low ubiquitination")} 
    else if(cna[k,1] > cna[k,2]) { group[k] = "High ubiquitination"}
    else { group[k] = NA} # paste0("Intermediate ubiquitination")
  }
  # Bind the grouping to the matrix.
  cna$Group = as.factor(group)
  # Keep only complete cases.
  cna = cna[complete.cases(cna),]
  rownames(cna) = as.character(base::gsub("\\.", "-", rownames(cna)))

  # Get the expression data. 
  if(file.exists("brca_tcga_rnaseq2.rds")) {
    exprs_TCGA = readRDS("brca_tcga_rnaseq2.rds")
  } else {
    exprs_TCGA = getTCGA(disease="BRCA", data.type="RNASeq2")$dat
    # Save the expression data for later use. 
    saveRDS(exprs_TCGA, "brca_tcga_rnaseq2.rds") 
  }
  # Log-transform.
  exprs = apply(exprs_TCGA, MARGIN = 1, function(x) log2(x+1))
  exprs = t(exprs) # The heck?? Why did apply() flip it??
  # Keep only the primary tumors. 
  type = sapply(colnames(exprs), function(s) unlist(strsplit(s, "-"))[4]) # Split the sample name up by dashes. The sample type is the fourth number (e.g. "01A") in the sample name.
  exprs =  exprs[, base::grep("^01", type)] # The primary tumors are the ones whose type has "01" in it. 
  # Shorten the sample names so we can match them with those in the cna data frame. 
  colnames(exprs) = substr(colnames(exprs), 1, 15)
  # Make sure it matches the cna matrix.
  common_samples = intersect(colnames(exprs), rownames(cna))
  exprs = exprs[,colnames(exprs) %in% common_samples]
  cna = cna[rownames(cna) %in% common_samples,]
  exprs = exprs[,order(colnames(exprs))]
  cna = cna[order(rownames(cna)),]
  identical(colnames(exprs), rownames(cna))

  # Perform differential expression analysis using limma.
  # Create the design matrix. 
  # Set the high-ubiquitination signature as the baseline (=1.) For the low-vs-high-ubiquitination column, 0 = high ubiquitination (i.e. no change from the baseline) and 1 = low ubiquitination (i.e. a change from the baseline.)
  design = cbind(HighUb=1, LowVsHighUb=ifelse(cna$Group=="High ubiquitination", 0, 1))
  # design = cbind(NeutUb=1, LowVsNeutUb=ifelse(cna$Group=="Intermediate ubiquitination", 0, 1))

  # Fit the model.
  fit = lmFit(exprs, design)
  fit = eBayes(fit)
  topTable(fit)
  limmares = topTable(fit, coef = 2, n = Inf, sort ="p")
  limmafc = limmares$logFC

  # Load in and process the downloaded CMap drug rank matrix.
  cmap_ref_profiles = get.cmap.ref(cmap.data.path = 'rankMatrix.txt', probe.to.genes = probe.to.genes, drug.info = drug.info)

  # Generate the data set from the limma results.
  # From the DrInsight R package vignette (https://cran.r-project.org/web/packages/DrInsight/vignettes/my-vignette.html): "The parameter query.data takes a matrix containing two columns: geneSymbol and score. The scores in ‘score’ column can be any numeric scores from statistical test of differential analysis between query and control phenotype. (Note: the column names of query.data matrix must be ‘geneSymbol’ and ‘score’)."
  # Following the DrInsight R package vignette, we will be using the t-statistic from limma as the score. 
  query_data = data.frame(geneSymbol=rownames(limmares), score=limmares$t)

  # Dr. Insight drug identification.
  drug_ident_res_neg = drug.ident(query.data = query_data, cmap.ref.profiles = cmap_ref_profiles, repurposing.unit = "treatment", connectivity = "negative") # To identify drugs that will produce the desired perturbations.
  drug_ident_res_pos = drug.ident(query.data = query_data, cmap.ref.profiles = cmap_ref_profiles, repurposing.unit = "treatment", connectivity = "positive") # To identify drugs that will produce the OPPOSITE of the desired perturbations.

  # select the drug p value table
  drug_pvals_neg = drug_ident_res_neg$drug.pvals
  drug_pvals_pos = drug_ident_res_pos$drug.pvals

  # Format the tables. 
  # Negative.
  drug_pvals_neg_formatted = tidyr::separate(data = drug_pvals_neg, col = drug, into = c("Drug", "CellLine"), sep = "_")
  colnames(drug_pvals_neg_formatted) = c("Drug", "Cell line", "p-value", "FDR")
  drug_pvals_neg_formatted = drug_pvals_neg_formatted[drug_pvals_neg_formatted$FDR < 0.05,]
  drug_pvals_neg_formatted$`p-value` = signif(drug_pvals_neg_formatted$`p-value`, 2)
  drug_pvals_neg_formatted$FDR = signif(drug_pvals_neg_formatted$FDR, 2)
  ft_neg = autofit(flextable(drug_pvals_neg_formatted))
  # Positive.
  drug_pvals_pos_formatted = tidyr::separate(data = drug_pvals_pos, col = drug, into = c("Drug", "CellLine"), sep = "_")
  colnames(drug_pvals_pos_formatted) = c("Drug", "Cell line", "p-value", "FDR")
  drug_pvals_pos_formatted = drug_pvals_pos_formatted[drug_pvals_pos_formatted$FDR < 0.05,]
  drug_pvals_pos_formatted$`p-value` = signif(drug_pvals_pos_formatted$`p-value`, 2)
  drug_pvals_pos_formatted$FDR = signif(drug_pvals_pos_formatted$FDR, 2)
  ft_pos = autofit(flextable(drug_pvals_pos_formatted))

  # Save the table. 
  saveRDS(ft_neg, "Table_4_neg.rds")
  saveRDS(ft_pos, "Table_4_pos.rds")
  
} else {
  ft_neg = readRDS("Table_4_neg.rds")
  ft_pos = readRDS("Table_4_pos.rds")
}

```

```{r echo = FALSE, eval = FALSE, fig.margin = TRUE, fig.cap = ""}
ft_neg
ft_pos
```

### Figure 8
Cell-line ER status: https://cancerres.aacrjournals.org/content/canres/suppl/2008/07/28/68.15.6084.DC1/tables_1-3.pdf

```{r echo=FALSE, eval=TRUE, include=FALSE}
cell_lines = c("MCF7", "T47D", "BT483", "ZR751", "ZR75b", "ZR7530", "MDAMB134VI", "MDAMB415", "CAMA1", "LY2", "600MPE", "HCC70", "BT474", "MDAMB361", "UACC812", "UACC893")
drugs = c("Tamoxifen", "Trichostatin A", "Rapamycin", "LY-294002", "Fulvestrant", "Wortmannin", "Tanespimycin", "Vorinostat") # , "Leflunomide", "Parthenolide", "Phenformin", "Piperlongumine", "Pyrimethamine", "Staurosporine", "Tanespimycin", "Thapsigargin", "Tretinoin"
drugs_negative = c("Tamoxifen", "Trichostatin A", "Rapamycin", "LY-294002", "Fulvestrant", "Wortmannin", "Tanespimycin", "Vorinostat")
# Drugs in GDSC: Tamoxifen, Trichostatin A, Rapamycin, Fulvestrant, Tanespimycin, Vorinostat.
drugs_positive_chemo = c("Paclitaxel", "Vinblastine",  "Carmustine",  "Mitoxantrone", "Camptothecin", "Doxorubicin", "Irinotecan", "Methotrexate", "Etoposide")
drugs_positive_targeted = c("Imatinib", "Gefitinib")
drugs_positive_other = c("Leflunomide", "Parthenolide", "Phenformin", "Piperlongumine", "Pyrimethamine", "Staurosporine", "Thapsigargin", "Tretinoin")
drugs_positive = union(union(drugs_positive_chemo, drugs_positive_targeted), drugs_positive_other)

# Fetch the GDSC cell-line data from cBioPortal. 
## Set the genes for the signature.
genes_sig = c("USP10", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

# Get the name of the cancer study in cBioPortal.
#View(getCancerStudies(mycgds))
# We want either ccle_broad_2019 or cellline_ccle_broad.
cancer_study = "ccle_broad_2019"
# Get the name of the case list.
#View(getCaseLists(mycgds, cancer_study))
# We want ccle_broad_2019_cna.
case_name = "ccle_broad_2019_cna"
# Get the name of the genetic profile.
#View(getGeneticProfiles(mycgds, cancer_study))
# We want ccle_broad_2019_cna.
genetic_profile = "ccle_broad_2019_cna"

# Get the CNA data.
cna_subset = getProfileData(mycgds, genes_sig, genetic_profile, case_name)
# Rename the row names.
cna_subset$CellLine = rownames(cna_subset) %>% regexPipes::gsub("\\_.+", "")
# Throw out all the samples that don't have expression data.
cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
# Classify cell lines into ubiquitination groups.
cna_subset = cna_subset %>% filter(CellLine %in% cell_lines)
group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
for(k in 1:nrow(cna_subset)) {
  if(cna_subset[k,1] < cna_subset[k,2]) { group[k] = paste0("Low ubiquitination")} 
  else if(cna_subset[k,1] > cna_subset[k,2]) { group[k] = paste0("High ubiquitination")}
  else { group[k] = "Intermediate ubiquitination"}
}
# Bind the grouping to the matrix.
cna_subset$Group = as.factor(group)
# Get only the low-ubiquitination cell lines.
low_ub_cell_lines = cna_subset %>% filter(Group=="Low ubiquitination") %>% dplyr::select(CellLine) %>% unlist()

# Load the drug-response data.
drug_response_1 = as.data.frame(read_excel("GDSC1_fitted_dose_response_25Feb20.xlsx"))
drug_response_2 = as.data.frame(read_excel("GDSC2_fitted_dose_response_25Feb20.xlsx"))
drug_response = rbind(drug_response_1, drug_response_2) %>% filter(TCGA_DESC=="BRCA")
drug_response$CELL_LINE_NAME = drug_response$CELL_LINE_NAME %>% regexPipes::gsub("\\-", "")
# Convert ln(IC50) to -log10(IC50).
drug_response$pIC50 = -log10(exp(drug_response$LN_IC50))

plots = list()
title_width = 50
# Figure 8a: compare the recommended drugs to the standard-of-care, tamoxifen.
# Keep only the recommended drugs. 
drug_response_recommended = drug_response %>% filter(DRUG_NAME %in% drugs)
drug_response_recommended$DRUG_NAME = drug_response_recommended$DRUG_NAME %>% as.factor %>% factor(levels = c("Tamoxifen", "Fulvestrant", "Rapamycin", "Tanespimycin", "Trichostatin A", "Vorinostat")) # , "Leflunomide", "Parthenolide", "Phenformin", "Piperlongumine", "Pyrimethamine", "Staurosporine", "Thapsigargin", "Tretinoin"

# Use ggplot2 to create a boxplot. 
# Classify the drugs as either tamoxifen or not. 
drug_response_recommended$DRUG_GROUP = ifelse(drug_response_recommended$DRUG_NAME=="Tamoxifen", "Tamoxifen", "Recommended drugs")
title_8a = "Comparison of efficacy of CMAP-recommended drugs vs tamoxifen in ER-positive BC cell lines"
plots[["Fig8a"]] = list()
plots[["Fig8a"]][["Fig8a"]] = ggplot(data=drug_response_recommended, aes(x=DRUG_NAME, y=pIC50, fill=DRUG_GROUP)) +
    geom_boxplot(lwd = boxplot_lwd) + # geom_violin() #, show.legend = FALSE https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
    guides(fill=guide_legend(title="Drug")) + 
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins, hjust = 0.5),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size)) +
    scale_fill_manual(values=c("#F0E685FF", "#749B58FF")) +
    labs(title=paste(strwrap(title_8a, width = title_width), collapse = "\n"), x="Drug", y=" ") + 
    ylim(NA, 4) + 
    #stat_compare_means(comparisons = comparisons,
    #                   label =  "p.signif", 
    #                   label.x = 1.5, 
    #                   label.y = 3.6,
    #                   bracket.size = 0,
    #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
     #                                     symbols = c("****", "***", "**", "*", " ")), 
    #                 size = p_value_size ) +
    stat_compare_means(method = "anova", 
                       label.y = 3.5,
                       size = p_value_size) + # Add global p-value
    stat_compare_means(aes(label = ..p.signif..),
                  method = "t.test", 
                  ref.group = "Tamoxifen",
                  label.y = 3.0,
                  size = p_value_size)

# Save the ANOVA results. 
# Perform ANOVA.
res_aov = aov(LN_IC50 ~ DRUG_NAME, data = drug_response_recommended)
# Summary of the analysis.
aov_stats = summary(res_aov)
# Perform post-hoc tests. 
library(DescTools)
dunnett_test = DunnettTest(x = drug_response_recommended$LN_IC50, g = drug_response_recommended$DRUG_NAME)
#TukeyHSD(res_aov)
stats = list()
stats[["ANOVA"]] = aov_stats
stats[["Dunnett"]] = dunnett_test
saveRDS(stats, file = "Figure8_statistics.rds")
saveRDS(plots, file = "Figure8_plot.rds")

# Combine the plots into a single figure.
figure_8a = ggpubr::ggarrange(plotlist = plots[["Fig8a"]], 
                   nrow = 1, 
                   ncol = 1, 
                   labels = " ",
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = FALSE)
figure_8a = annotate_figure(figure_8a,
                           left = text_grob(expression(-log[10](IC[50])), rot = 90, size = common_axis_text_y_size),
                           fig.lab = " ",) # expression(-log[10](IC[50])) 

plots_Fig8 = list()
plots_Fig8$Fig8a = figure_8a
#plots_Fig8$Fig8b = figure_8b
figure_8 = ggpubr::ggarrange(plotlist = plots_Fig8,
                             nrow = 1,
                             ncol = 1,
                             hjust = c(-0.5, -1, -0.5, -1, -1.5))

# Save to EPS.
setEPS()
postscript("Figure_8.eps", width = 20, height = 20)
figure_8
dev.off()
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 15, fig.margin = TRUE, fig.cap = ""}
#plotlist = list(figure_8a)
#grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)

#plotlist = list(figure_8b)
#grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)

plotlist = list(figure_8)
grid.arrange(grobs = lapply(plots_Fig8, "+", margin), nrow = 1)
```

## Supplemental information
### Figure S2 - Association of CNA with p27 levels (cf. Figure 2)
```{r echo = FALSE, eval = TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

# Divide the patients into the four molecular subtypes as listed here: https://ww5.komen.org/BreastCancer/SubtypesofBreastCancer.html. 
# On using .$ to extract column as vector: https://stackoverflow.com/a/27160317 
clin_dat = read.delim("brca_tcga_clinical_data.tsv", stringsAsFactors = F)

# For each gene in single_genes, 1) get the copy numbers of that gene and the protein levels of p27 for all the samples in the subtype, 2) split the samples into two groups by copy-number status, and 3) compare the p27 levels between the two groups. 
# EDIT: As of 08/06/2020, the only single gene we'll be looking at is SKP2; i.e., we won't divide samples into groups based on USP10/USP13/FZR1 alone. This will allow us to name the groups as "High SKP2" or "Low SKP2," rather than "High ubiquitination" or "Low ubiquitination" and color code these groups separately. 
single_genes = c("SKP2", "USP10", "USP13", "FZR1")
downstream_gene = "CDKN1B"
dat_tables = list()
for(j in 1:length(single_genes)) {
    gene = single_genes[j]
    
    # Get all the samples that have target protein expression.
    #prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_rppa"))[,1,drop=FALSE]
    prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_all"))[,1,drop=FALSE]
    # Remove all samples that do not have expression data.
    prot = prot[complete.cases(prot),,drop=FALSE]
    # if(nrow(prot)==0) next # If no protein data, skip to the next gene.
    
    # Get the CNA data of the predictor genes.
    cna = getProfileData(mycgds, gene, paste0(tolower(cancer), "_tcga_gistic"), paste0(tolower(cancer), "_tcga_cna"))
    # Filter out incomplete cases. 
    # Remove incomplete cases (NAs or NaNs).
    cna = cna[complete.cases(cna),,drop=FALSE]
    if(nrow(cna)==0) next # If no cna data, skip to the next gene.
    
    # Keep only the samples that have both gene and protein levels and meet the criteria for the current subtype. 
    good_samples = intersect(rownames(cna), rownames(prot))
    prot = prot[which(rownames(prot) %in% good_samples),,drop=FALSE]
    cna = cna[which(rownames(cna) %in% good_samples),,drop=FALSE]
    # if(nrow(prot)==0 | nrow(cna)==0) next 
    
    # Make sure the samples are in the same order in both the protein and gene copy number vectors.
    prot = prot[order(rownames(prot)),,drop=FALSE]
    cna = cna[order(rownames(cna)),,drop=FALSE]
    # identical(rownames(prot), rownames(cna))
    # Make sure the column names of the predictor genes are in alphabetical order.
    cna = cna[,order(colnames(cna), decreasing = FALSE),drop=FALSE]
    identical(rownames(cna), rownames(prot))
    
    subtypes = list()
    subtypes[["All"]] = clin_dat %>% .$Sample.ID
    subtypes[["Luminal"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive") %>% .$Sample.ID
    subtypes[["HER2"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Positive") %>% .$Sample.ID
#subtypes[["Non-TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive" | PR.status.by.ihc=="Positive" | IHC.HER2=="Positive") %>% .$Sample.ID
    subtypes[["TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Negative") %>% .$Sample.ID

    # For each subtype ... 
    plots = list()
    dat_tables[[gene]] = list()
    prot_name = dict[[downstream_gene]]
    for(subtype in names(subtypes)) {
      # Set up the data frame that will hold the data for all the cancers, for subtype "subtype". 
      dat = data.frame(Genes=character(), ProteinLevels=numeric(), Group=character())
      # Get the sample IDs for this subtype.
      sample_IDs = base::gsub("\\-", "\\.", subtypes[[subtype]])
      cna_subtype = cna[rownames(cna) %in% sample_IDs,,drop=FALSE]
      prot_subtype = prot[rownames(prot) %in% sample_IDs,,drop=FALSE]
      
      # Divide samples by copy number of the predictor genes.
      # Create the signature matrix.
      group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
      for(k in 1:nrow(cna_subtype)) {
        if(cna_subtype[k,1] < 0) { group[k] = paste0("Low ", gene)}
        else if(cna_subtype[k,1] > 0) {group[k] = paste0("High ", gene)}
        else { group[k] = paste0("Neutral ", gene)}
      }
      # Bind the grouping to the matrix.
      cna_subtype$Group = as.factor(group)
      # Keep only complete cases.
      cna_subtype = cna_subtype[complete.cases(cna_subtype),,drop=FALSE]
      prot_subtype = prot_subtype[rownames(prot_subtype) %in% rownames(cna_subtype),,drop=FALSE]
      # Re-order the levels of "Group": Low - Neutral - High.
      cna_subtype$Group = factor(cna_subtype$Group, levels = c(paste0('Low ', gene), paste0('Neutral ', gene), paste0('High ', gene)))
      if(identical(rownames(cna_subtype), rownames(prot_subtype))) {
        prot_subtype$Group = cna_subtype$Group
        
        dat = rbind(dat, data.frame(Genes=gene, ProteinLevels=prot_subtype[,1], Group=prot_subtype$Group))
      }
      
      # Save the counts.
      counts_table_final = rbind(counts_table_final, data.frame(
        Gene = gene,
        Subtype = subtype,
        Count = nrow(dat),
        stringsAsFactors = F
        ))
  
      # Save dat to the list.
      dat_tables[[gene]][[subtype]] = dat
      write.table(dat, file = paste0("FigureS2_cna_gene_", gene, "_subtype_", subtype, ".csv"), sep = ",", row.names = F, col.names = F)
      
      # Use ggplot2 to create a dotplot. 
      plots[[subtype]] = ggplot(data=dat, aes(x=Genes, y=ProteinLevels, fill=Group, group=interaction(Group, Genes))) + # Each separate boxplot will be the "intersection" (i.e. interaction) of a gene (or gene signature) x group.
        geom_boxplot(lwd = boxplot_lwd) + # geom_violin() # https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
        theme_classic() + 
        theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
              axis.line = element_line(size = axis_line_size),
              axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
              axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
              axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
              axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
              legend.title = element_text(size = legend_title_size),
              legend.text = element_text(size = legend_text_size)) +
        scale_fill_manual(values=c("#0073C2FF", "#868686FF", "#EFC000FF")) + #scale_fill_jco() +
        labs(title=subtype, x=" ", y=" ") #+ 
        #stat_compare_means(label =  "p.signif", 
        #                   label.x = 1.5, 
        #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
        #                                      symbols = c("****", "***", "**", "*", " ")), 
        #                   size = p_value_size )  
    }
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots, 
                               nrow = 2, 
                               ncol = 2, 
                               #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                               hjust = c(-0.5, -1, -0.5, -1, -1.5),
                               font.label = list(size = 20),
                               common.legend = TRUE, 
                               legend = "right")
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    figure_S2_i = annotate_figure(figure,
                                  top = text_grob(paste(prot_name, "protein levels between", gene, "copy number groups"), size = overall_title_size),
                                  left = text_grob(paste(prot_name, "protein levels"), rot = 90, size = common_axis_text_y_size),
                                  bottom = text_grob(paste("Gene"), size = common_axis_text_x_size),
                                  fig.lab = letters[j],
                                  fig.lab.size = figure_label_size,
                                  fig.lab.face = figure_label_face
                                  )
    # show_col(pal_jco()(4))
    
    # Save to EPS.
    setEPS()
    postscript(paste0("Figure_S2", letters[j], ".eps"), width = 20, height = 20)
    print(figure_S2_i)
    dev.off()
}

# Save the counts.
write.table(counts_table_final, "FigureS2_counts.csv", sep = ",", quote = F, row.names = F)

# Statistics.
# http://www.sthda.com/english/wiki/one-way-anova-test-in-r
stats = list()
for(gene_set in names(dat_tables)) {
  stats[[gene_set]] = list()
  
  for(subtype in names(subtypes)) {
    stats[[gene_set]][[subtype]] = list()
    dat_sub = dat_tables[[gene_set]][[subtype]] %>% droplevels()
    
    # Compute summary statistics.
    summary_stats = dat_sub %>% group_by(Group) %>%
      summarise(
        Count = n(),
        Mean = mean(ProteinLevels, na.rm = TRUE),
        SD = sd(ProteinLevels, na.rm = TRUE)
        )
    # Add a sum of counts. 
    summary_stats = rbind(summary_stats, data.frame(Group = "Total", Count = sum(summary_stats$Count), Mean = NA, SD = NA))
    # Save to stats list.
    stats[[gene_set]][[subtype]][["Summary"]] = summary_stats
    
    # Perform ANOVA.
    res_aov = aov(ProteinLevels ~ Group, data = dat_sub)
    # Summary of the analysis.
    aov_stats = summary(res_aov)
    # Save to stats list.
    stats[[gene_set]][[subtype]][["Inferential"]] = aov_stats
  }
}
saveRDS(stats, file = "FigureS2_statistics.rds")
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
plotlist = list(figure_S2)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```


```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
figure_S2_i
```


### Figure S3 - Prognostic impact of CNA of signature genes individually (cf. Figure 4)
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

plots = list()
stats = list()
for(sig_gene in genes_sig) {
  plots[[sig_gene]] = list()
  
  # Get the expression data.
  cna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_cna", "brca_metabric_cna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
  
  # Throw out all the samples that don't have expression data.
  cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
  # Subset to include only the patients with no chemo, hormone, or radiation therapy. 
  # clin_sig = clin_sig[clin_sig$CHEMOTHERAPY=="NO" & clin_sig$HORMONE_THERAPY=="NO" & clin_sig$RADIO_THERAPY=="NO",,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients.
  overlap = intersect(rownames(clin_sig), rownames(cna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  cna_subset = cna_subset[order(rownames(cna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(cna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  # Do survival analysis for each subtype. 
  for(subtype in names(subtypes)) {
    # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
    ## Divide up the samples by copy number of the genes in the signature.
    cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < 0) { cna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(cna_subset_subtype[k,1] > 0) { cna_subset_group[k] = paste0("High ", sig_gene)} 
      else { cna_subset_group[k] = paste0("Neutral ", sig_gene) }
    }
    cna_subset_subtype$Group = cna_subset_group
    # Remove NAs.
    cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
    identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))
      
    # Re-order the levels of "Group": Low - Neutral - High.
    cna_subset_subtype$Group = factor(cna_subset_subtype$Group, levels = c(paste0('Low ', sig_gene), paste0('Neutral ', sig_gene), paste0('High ', sig_gene)))
    
    ## Perform survival analysis.
    # Create the survival matrix.
    surv_matrix_sig = clin_sig_subtype[,which(colnames(clin_sig_subtype) %in% c("OS_MONTHS", "VITAL_STATUS"))]
    surv_matrix_sig$VITAL_STATUS = as.numeric(ifelse(surv_matrix_sig$VITAL_STATUS=="Died of Disease", 1, 0))
    identical(rownames(surv_matrix_sig), rownames(cna_subset_subtype))
    surv_matrix_sig$Group = cna_subset_subtype$Group # Make sure the clinical and expression samples are IDENTICAL (same samples in the same order) before doing this! 
    # Create the survival object.
    surv_obj_sig = survival::Surv(time = surv_matrix_sig$OS_MONTHS, event = surv_matrix_sig$VITAL_STATUS)
    # Fit the survival curves.
    surv_model_sig = survival::survfit(surv_obj_sig ~ Group, data = surv_matrix_sig)
    # Save the stats to list.
    stats[[sig_gene]][[subtype]] = surv_model_sig
    
    # Plot.
    ggsurv_cna_ub = ggsurvplot(surv_model_sig, 
                               data = surv_matrix_sig, 
                               font.tickslab = tick_label_size,
                               size = km_line_size, # Line size.
                               pval = TRUE,
                               pval.size = p_value_size,
                               conf.int = F, # Show confidence intervals for point estimates of survival curves.
                               break.time.by = 50,
                               title = subtype,
                               xlab = " ",   # Customize X axis label.
                               ylab = " ",
                               legend.title = "Group",
                               legend.labs = c(paste0('Low ', sig_gene), paste0('Neutral ', sig_gene), paste0('High ', sig_gene)),
                               risk.table = FALSE, # Set to TRUE to turn risk tables on. 
                               risk.table.y.text.col = T, # Colour risk table text annotations.
                               risk.table.y.text = FALSE, # Show bars instead of names in text annotations in legend of risk table)
                               palette = c("#0073C2FF", "#868686FF", "#EFC000FF") #, "#868686FF"
                               ) 
    
    # https://github.com/kassambara/survminer/issues/2
    ggsurv_cna_ub$plot = ggsurv_cna_ub$plot +
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size))
    #ggsurv_cna_ub$table = ggsurv_cna_ub$table + labs(title = "Number at risk") # Uncomment this line when turning risk tables back on. 
    
    # Save to the plots list. 
    plots[[sig_gene]][[subtype]] = ggsurv_cna_ub$plot
    # print(ggsurv_cna_ub)
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      Count = sum(surv_model_sig$n),
      stringsAsFactors = F
      ))
  }
}


# Arrange plots. 
for(i in 1:length(genes_sig)) {
  sig_gene = genes_sig[i]
  figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]], 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
  # Annotate arranged ggsurvplots with common labels. 
  figure_S3_i = annotate_figure(figure,
                               top = text_grob(paste0("Overall survival of breast-cancer patients by ", sig_gene, " copy number"), size = overall_title_size),
                               left = text_grob(paste("Survival probability"), rot = 90, size = common_axis_text_y_size),
                               bottom = text_grob(paste("Survival time (months)"), size = common_axis_text_x_size),
                               fig.lab = letters[i], 
                               fig.lab.size = figure_label_size,
                               fig.lab.face = figure_label_face
                               )
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S3", letters[i], ".eps"), width = 20, height = 20)
  print(figure_S3_i)
  dev.off()

}

# Save statistics to file. 
saveRDS(stats, file = "FigureS3_statistics.rds")
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
figure_S3_i
```

### Figure S4 - Prognostic impact of expression of signature genes individually (cf. Figure 4)
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

stats = list()
plots = list()
upper_quantile_sig = 0.50
lower_quantile_sig = 0.50
for(sig_gene in genes_sig) {
  plots[[sig_gene]] = list()
  
  # Get the expression data.
  mrna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_mrna", "brca_metabric_mrna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_mrna")
  
  # Throw out all the samples that don't have expression data.
  mrna_subset = mrna_subset[complete.cases(mrna_subset),,drop=FALSE]
  # Subset to include only the patients with no chemo, hormone, or radiation therapy. 
  # clin_sig = clin_sig[clin_sig$CHEMOTHERAPY=="NO" & clin_sig$HORMONE_THERAPY=="NO" & clin_sig$RADIO_THERAPY=="NO",,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients.
  overlap = intersect(rownames(clin_sig), rownames(mrna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  mrna_subset = mrna_subset[rownames(mrna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  mrna_subset = mrna_subset[order(rownames(mrna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(mrna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  # Do survival analysis for each subtype. 
  for(subtype in names(subtypes)) {
    # Subset both mrna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    mrna_subset_subtype = mrna_subset[rownames(mrna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
    ## Divide up the samples by copy number of the genes in the signature.
    mrna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    for(k in 1:nrow(mrna_subset_subtype)) {
      if(mrna_subset_subtype[k,1] <= quantile(mrna_subset_subtype[,1], lower_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(mrna_subset_subtype[k,1] > quantile(mrna_subset_subtype[,1], upper_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("High ", sig_gene)} 
      else { mrna_subset_group[k] = NA }
    }
    mrna_subset_subtype$Group = mrna_subset_group
    # Remove NAs.
    mrna_subset_subtype = mrna_subset_subtype[complete.cases(mrna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(mrna_subset_subtype)),]
    identical(rownames(mrna_subset_subtype), rownames(clin_sig_subtype))
    
    # Re-order the levels of "Group": Low - High.
    mrna_subset_subtype$Group = factor(mrna_subset_subtype$Group, levels = c(paste0('Low ', sig_gene), paste0('High ', sig_gene)))
    
    ## Perform survival analysis.
    # Create the survival matrix.
    surv_matrix_sig = clin_sig_subtype[,which(colnames(clin_sig_subtype) %in% c("OS_MONTHS", "VITAL_STATUS"))]
    surv_matrix_sig$VITAL_STATUS = as.numeric(ifelse(surv_matrix_sig$VITAL_STATUS=="Died of Disease", 1, 0))
    identical(rownames(surv_matrix_sig), rownames(mrna_subset_subtype))
    surv_matrix_sig$Group = mrna_subset_subtype$Group # Make sure the clinical and expression samples are IDENTICAL (same samples in the same order) before doing this! 
    # Create the survival object.
    surv_obj_sig = survival::Surv(time = surv_matrix_sig$OS_MONTHS, event = surv_matrix_sig$VITAL_STATUS)
    # Fit the survival curves.
    surv_model_sig = survival::survfit(surv_obj_sig ~ Group, data = surv_matrix_sig)
    # Save the stats to list.
    stats[[sig_gene]][[subtype]] = surv_model_sig
    
    # Plot.
    ggsurv_mrna_ub = ggsurvplot(surv_model_sig, 
                               data = surv_matrix_sig, 
                               font.tickslab = tick_label_size,
                               size = km_line_size, # Line size.
                               pval = TRUE,
                               pval.size = p_value_size,
                               conf.int = F, # Show confidence intervals for point estimates of survival curves.
                               break.time.by = 50,
                               title = subtype,
                               xlab = " ",   # Customize X axis label.
                               ylab = " ",
                               legend.title = "Group",
                               legend.labs = c(paste0('Low ', sig_gene), paste0('High ', sig_gene)),
                               risk.table = FALSE, # Set to TRUE to turn risk tables on. 
                               risk.table.y.text.col = T, # Colour risk table text annotations.
                               risk.table.y.text = FALSE, # Show bars instead of names in text annotations in legend of risk table)
                               palette = c("#0073C2FF", "#EFC000FF") #, "#868686FF"
                               ) 
    
    # https://github.com/kassambara/survminer/issues/2
    ggsurv_mrna_ub$plot = ggsurv_mrna_ub$plot +
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size))
    #ggsurv_mrna_ub$table = ggsurv_mrna_ub$table + labs(title = "Number at risk") # Uncomment this line when turning risk tables back on. 
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      Count = sum(surv_model_sig$n),
      stringsAsFactors = F
      ))
    
    # Save to the plots list. 
    plots[[sig_gene]][[subtype]] = ggsurv_mrna_ub$plot
    # print(ggsurv_mrna_ub)
  }
}


# Arrange plots. 
for(i in 1:length(genes_sig)) {
  sig_gene = genes_sig[i]
  figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]], 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
  # Annotate arranged ggsurvplots with common labels. 
  figure_S4_i = annotate_figure(figure,
                               top = text_grob(paste0("Overall survival of breast-cancer patients by ", sig_gene, " expression"), size = overall_title_size),
                               left = text_grob(paste("Survival probability"), rot = 90, size = common_axis_text_y_size),
                               bottom = text_grob(paste("Survival time (months)"), size = common_axis_text_x_size),
                               fig.lab = letters[i], 
                               fig.lab.size = figure_label_size,
                               fig.lab.face = figure_label_face
                               )
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S4", letters[i], ".eps"), width = 20, height = 20)
  print(figure_S4_i)
  dev.off()

}

# Save statistics to file. 
saveRDS(stats, file = "FigureS4_statistics.rds")
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
figure
```

### Figure S5 and S6 - Clinical association of CNA of signature genes individually (cf. Figures 5 and 6)
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  ClinicalVariable = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

plots = list()
results = list()
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  plots[[sig_gene]] = list()
  results[[sig_gene]] = list()
  
  # Get the expression data.
  cna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_cna", "brca_metabric_cna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  # Throw out all the samples that don't have cna data.
  cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients.
  overlap = intersect(rownames(clin_sig), rownames(cna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  cna_subset = cna_subset[order(rownames(cna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(cna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  for(subtype in names(subtypes)) {
    # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
    ## Divide up the samples by copy number of the genes in the signature.
    cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < 0) { cna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(cna_subset_subtype[k,1] > 0) { cna_subset_group[k] = paste0("High ", sig_gene)} 
      else { cna_subset_group[k] = paste0("Neutral ", sig_gene) }
    }
    cna_subset_subtype$Group = cna_subset_group
    # Remove NAs.
    cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
    identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))
    
    # Create a table holding, for each sample, 1) SKP2 ubiquitination signature, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
    other_clin_features = data.frame(UbSig=cna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
    
    # Stage
    # Use chi-squared test. 
    confusion_stage = table(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
    chisq_test_stage = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
    # Graph.
    proptable_stage = prop.table(confusion_stage,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
    melted_stage = melt(proptable_stage)
    colnames(melted_stage) = c("UbSignature", "Stage", "Proportion")
    melted_stage$Stage = as.factor(melted_stage$Stage)
    # Get the y-value for the p-value annotation. 
    pval_y = min(max(melted_stage$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
    pval = as.character(signif(chisq_test_stage$p.value, 2))
    # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
    # Re-order the levels of "UbSignature": Low - Neutral - High.
    melted_stage$UbSignature = factor(melted_stage$UbSignature, levels = c(paste0("Low ", sig_gene), paste0("Neutral ", sig_gene), paste0("High ", sig_gene)))
    
    stage_plot = ggplot(melted_stage, aes(x = UbSignature, y = Proportion, fill = Stage)) +
      geom_bar(stat = "identity", position=position_dodge()) + #color="black",
      #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
      annotate("text", 
               x = 2, 
               y = pval_y, 
               size = p_value_size,
               label = paste("p==", pval), #as.expression(bquote("Chi-square"~italic(p)==bold(.(pval))))
               parse = TRUE) + # Annotate with p-value. 
      # https://www.r-bloggers.com/math-notation-for-r-plot-titles-expression-and-bquote/
      # https://r.789695.n4.nabble.com/lattice-plotmath-how-to-get-a-variable-in-bold-face-td3569599.html
      scale_fill_manual(breaks = c('0', '1', '2', '3', '4'),
                        values=pal_npg()(5)) +
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
      #scale_size(range = c(0, 10)) +
      labs(x=" ", y=" ", title=paste0(subtype))
    
    print(paste0("The p-value of the chi-squared test on the distribution of tumor stage by ubiquitination signature in subtype ", subtype, " is ", chisq_test_stage$p.value))
    
    # Grade
    # Use chi-squared test.
    confusion_grade = table(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
    chisq_test_grade = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
    # Graph.
    proptable_grade = prop.table(confusion_grade,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
    melted_grade = melt(proptable_grade)
    colnames(melted_grade) = c(paste0("SigGeneStatus"), "Grade", "Proportion")
    melted_grade$Grade = as.factor(melted_grade$Grade)
    # Get the y-value for the p-value annotation. 
    pval_y = min(max(melted_grade$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
    pval = as.character(signif(chisq_test_grade$p.value, 2))
    # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
    # Re-order the factor levels of SigGeneStatus.
    melted_grade$SigGeneStatus = factor(melted_grade$SigGeneStatus, levels = c(paste0("Low ", sig_gene), paste0("Neutral ", sig_gene), paste0("High ", sig_gene)))
    
    # Plot.
    grade_plot = ggplot(melted_grade, aes(x = SigGeneStatus, y = Proportion, fill = Grade)) +
      geom_bar(stat = "identity", position=position_dodge()) + 
      #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
      annotate("text", 
               x = 2, 
               y = pval_y, 
               size = p_value_size,
               label =  paste("p==", pval), # as.expression(bquote("Chi-square"~italic(p)==bold(.(pval)))) 
               parse = TRUE) + # Annotate with p-value. 
      #scale_x_continuous(breaks=seq(1, 3, 1)) +
      scale_fill_manual(breaks = c('1', '2', '3'),
                        values=pal_npg()(3)) + 
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
      #scale_size(range = c(0, 10)) + 
      labs(x=" ", y=" ", title=paste0(subtype))
    
    print(paste0("The p-value of the chi-squared test on the distribution of tumor grade by ", sig_gene, " copy number in subtype ", subtype, " is ", chisq_test_grade$p.value))
  
    # Number of lymph nodes.
    # https://data.princeton.edu/wws509/r/overdispersion
    # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
    # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
    # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
    mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
    p_vals = coef(summary(mnb))[,4]
    
    # Add plots to plots list.
    plots[[sig_gene]][["Stage"]][[subtype]] = stage_plot
    plots[[sig_gene]][["Grade"]][[subtype]] = grade_plot
    # Add statistical test results to list. 
    results[[sig_gene]][["Stage"]][["Confusion"]][[subtype]] = confusion_stage
    results[[sig_gene]][["Stage"]][["Test"]][[subtype]] = chisq_test_stage
    results[[sig_gene]][["Grade"]][["Confusion"]][[subtype]] = confusion_grade
    results[[sig_gene]][["Grade"]][["Test"]][[subtype]] = chisq_test_grade
    results[[sig_gene]][["Nodes"]][["Test"]][[subtype]] = summary(mnb)
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      ClinicalVariable = "Stage",
      Count = sum(confusion_stage),
      stringsAsFactors = F
      ))
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      ClinicalVariable = "Grade",
      Count = sum(confusion_grade),
      stringsAsFactors = F
      ))
  }
  saveRDS(results, file = paste0("FigureS5_S6_", sig_gene, "_statistics.rds"))
  saveRDS(plots, file = paste0("FigureS5_S6_", sig_gene, "_plots.rds"))
    
  annotated_fig_list = list()
  for(variable in c("Stage", "Grade")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]][[variable]], 
                               nrow = 2, 
                               ncol = 2, 
                               #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                               #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                               font.label = list(size = 20),
                               common.legend = TRUE, 
                               legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                       top = text_grob(paste("Association of", sig_gene, "copy number with", str_to_lower(full_variable)), size = overall_title_size),
                                       left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                       bottom = text_grob(paste(sig_gene, "copy number group"), size = common_axis_text_x_size),
                                       fig.lab = letters[j], #paste0(str_to_lower(letters[i])), 
                                       fig.lab.size = figure_label_size,
                                       fig.lab.face = figure_label_face
                                       )
    annotated_fig_list[[variable]] = figure_annotated
    
  }
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S5", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Stage)
  dev.off()
  
  setEPS()
  postscript(paste0("Figure_S6", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Grade)
  dev.off()
}

```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Stage)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Grade)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Figure S7 and S8 - Clinical association of expression of signature genes individually (cf. Figures 5 and 6)
```{r echo = FALSE, eval=TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  ClinicalVariable = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

plots = list()
results = list()
upper_quantile_sig = 0.50
lower_quantile_sig = 0.50
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  plots[[sig_gene]] = list()
  results[[sig_gene]] = list()
  
  # Get the expression data.
  mrna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_mrna", "brca_metabric_mrna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_mrna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  # Throw out all the samples that don't have mrna data.
  mrna_subset = mrna_subset[complete.cases(mrna_subset),,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients.
  overlap = intersect(rownames(clin_sig), rownames(mrna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  mrna_subset = mrna_subset[rownames(mrna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  mrna_subset = mrna_subset[order(rownames(mrna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(mrna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  for(subtype in names(subtypes)) {
    # Subset both mrna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    mrna_subset_subtype = mrna_subset[rownames(mrna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
    ## Divide up the samples by copy number of the genes in the signature.
    mrna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    for(k in 1:nrow(mrna_subset_subtype)) {
      if(mrna_subset_subtype[k,1] <= quantile(mrna_subset_subtype[,1], lower_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(mrna_subset_subtype[k,1] > quantile(mrna_subset_subtype[,1], upper_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("High ", sig_gene)} 
      else { mrna_subset_group[k] = NA }
    }
    mrna_subset_subtype$Group = mrna_subset_group
    # Remove NAs.
    mrna_subset_subtype = mrna_subset_subtype[complete.cases(mrna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(mrna_subset_subtype)),]
    identical(rownames(mrna_subset_subtype), rownames(clin_sig_subtype))
    
    # Re-order the levels of "Group": Low - High.
    mrna_subset_subtype$Group = factor(mrna_subset_subtype$Group, levels = c(paste0('Low ', sig_gene), paste0('High ', sig_gene)))
    
    # Create a table holding, for each sample, 1) SKP2 ubiquitination signature, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
    other_clin_features = data.frame(UbSig=mrna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
    
    # Stage
    # Use chi-squared test. 
    confusion_stage = table(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
    chisq_test_stage = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Stage))
    # Graph.
    proptable_stage = prop.table(confusion_stage,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
    melted_stage = melt(proptable_stage)
    colnames(melted_stage) = c("UbSignature", "Stage", "Proportion")
    melted_stage$Stage = as.factor(melted_stage$Stage)
    # Get the y-value for the p-value annotation. 
    pval_y = min(max(melted_stage$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
    pval = as.character(signif(chisq_test_stage$p.value, 2))
    # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
    # Re-order the levels of "UbSignature": Low - High.
    melted_stage$UbSignature = factor(melted_stage$UbSignature, levels = c(paste0("Low ", sig_gene), paste0("High ", sig_gene)))
    
    stage_plot = ggplot(melted_stage, aes(x = UbSignature, y = Proportion, fill = Stage)) +
      geom_bar(stat = "identity", position=position_dodge()) + #color="black",
      #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
      annotate("text", 
               x = 2, 
               y = pval_y, 
               size = p_value_size,
               label = paste("p==", pval), #as.expression(bquote("Chi-square"~italic(p)==bold(.(pval))))
               parse = TRUE) + # Annotate with p-value. 
      # https://www.r-bloggers.com/math-notation-for-r-plot-titles-expression-and-bquote/
      # https://r.789695.n4.nabble.com/lattice-plotmath-how-to-get-a-variable-in-bold-face-td3569599.html
      scale_fill_manual(breaks = c('0', '1', '2', '3', '4'),
                        values=pal_npg()(5)) +
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
      #scale_size(range = c(0, 10)) +
      labs(x=" ", y=" ", title=paste0(subtype))
    
    print(paste0("The p-value of the chi-squared test on the distribution of tumor stage by ", sig_gene, " expression in subtype ", subtype, " is ", chisq_test_stage$p.value))
    
    # Grade
    # Use chi-squared test.
    confusion_grade = table(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
    chisq_test_grade = chisq.test(other_clin_features$UbSig, as.factor(other_clin_features$Grade))
    # Graph.
    proptable_grade = prop.table(confusion_grade,1) # 1 = rows; rows = ubiquitination groups. For each ubiquitination group, we want the proportion of patients of each grade. 
    melted_grade = melt(proptable_grade)
    colnames(melted_grade) = c(paste0("SigGeneStatus"), "Grade", "Proportion")
    melted_grade$Grade = as.factor(melted_grade$Grade)
    # Get the y-value for the p-value annotation. 
    pval_y = min(max(melted_grade$Proportion) + 0.05, 1)#max(melted_stage$Proportion) + (1 - max(melted_stage$Proportion))/2
    pval = as.character(signif(chisq_test_grade$p.value, 2))
    # https://stackoverflow.com/questions/52661393/how-to-italicise-part-of-correlation-coefficient-annotation-in-ggplot2
    # Re-order the factor levels of SigGeneStatus.
    melted_grade$SigGeneStatus = factor(melted_grade$SigGeneStatus, levels = c(paste0("Low ", sig_gene), paste0("High ", sig_gene)))
    
    # Plot.
    grade_plot = ggplot(melted_grade, aes(x = SigGeneStatus, y = Proportion, fill = Grade)) +
      geom_bar(stat = "identity", position=position_dodge()) + 
      #geom_point(aes(size=Number),shape=21, colour="#868686FF", fill="#868686FF") +
      annotate("text", 
               x = 2, 
               y = pval_y, 
               size = p_value_size,
               label =  paste("p==", pval), # as.expression(bquote("Chi-square"~italic(p)==bold(.(pval)))) 
               parse = TRUE) + # Annotate with p-value. 
      #scale_x_continuous(breaks=seq(1, 3, 1)) +
      scale_fill_manual(breaks = c('1', '2', '3'),
                        values=pal_npg()(3)) +
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
      #scale_size(range = c(0, 10)) + 
      labs(x=" ", y=" ", title=paste0(subtype))
    
    print(paste0("The p-value of the chi-squared test on the distribution of tumor grade by ", sig_gene, " expression in subtype ", subtype, " is ", chisq_test_grade$p.value))
  
    # Number of lymph nodes.
    # https://data.princeton.edu/wws509/r/overdispersion
    # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
    # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
    # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
    mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
    p_vals = coef(summary(mnb))[,4]
    
    # Add plots to plots list.
    plots[[sig_gene]][["Stage"]][[subtype]] = stage_plot
    plots[[sig_gene]][["Grade"]][[subtype]] = grade_plot
    # Add statistical test results to list. 
    results[[sig_gene]][["Stage"]][["Confusion"]][[subtype]] = confusion_stage
    results[[sig_gene]][["Stage"]][["Test"]][[subtype]] = chisq_test_stage
    results[[sig_gene]][["Grade"]][["Confusion"]][[subtype]] = confusion_grade
    results[[sig_gene]][["Grade"]][["Test"]][[subtype]] = chisq_test_grade
    results[[sig_gene]][["Nodes"]][["Test"]][[subtype]] = summary(mnb)
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      ClinicalVariable = "Stage",
      Count = sum(confusion_stage),
      stringsAsFactors = F
      ))
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      ClinicalVariable = "Grade",
      Count = sum(confusion_grade),
      stringsAsFactors = F
      ))
  }
  saveRDS(results, file = paste0("FigureS7_S8_", sig_gene, "_statistics.rds"))
  saveRDS(plots, file = paste0("FigureS7_S8", sig_gene, "_plots.rds"))
  
  # Get counts.
  counts_table = data.frame(
    Subtype = character(),
    Variable = character(),
    Count = integer()
    )
  
  for(variable in names(results[[sig_gene]])) {
    for(subtype in names(results[[sig_gene]][[variable]][["Confusion"]])) {
      count = results[[sig_gene]][[variable]][["Confusion"]][[subtype]] %>% sum
      
      counts_table = rbind(counts_table, data.frame(
        Subtype = subtype,
        Variable = variable,
        Count = count
      ))
    }
  }
  
  saveRDS(results, file = paste0("FigureS7_S8_", sig_gene, "_counts.rds"))
  
  annotated_fig_list = list()
  for(variable in c("Stage", "Grade")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]][[variable]], 
                               nrow = 2, 
                               ncol = 2, 
                               #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                               #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                               font.label = list(size = 20),
                               common.legend = TRUE, 
                               legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                       top = text_grob(paste("Association of", sig_gene, "expression with", str_to_lower(full_variable)), size = overall_title_size),
                                       left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                       bottom = text_grob(paste(sig_gene, "expression group"), size = common_axis_text_x_size),
                                       fig.lab = letters[j], #paste0(str_to_lower(letters[i])), 
                                       fig.lab.size = figure_label_size,
                                       fig.lab.face = figure_label_face
                                       )
    annotated_fig_list[[variable]] = figure_annotated
    
  }
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S7", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Stage)
  dev.off()
  
  setEPS()
  postscript(paste0("Figure_S8", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Grade)
  dev.off()
}

```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Stage)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

```{r echo = FALSE, eval = TRUE, fig.width = 23, fig.height = 23, fig.margin = TRUE, fig.cap = ""}
plotlist = list(annotated_fig_list$Grade)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Figure S9 - Clinical association of CNA of signature genes individually (cf. Figure 7)
```{r}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  
  # Get the expression data.
  cna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_cna", "brca_metabric_cna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  # Throw out all the samples that don't have cna data.
  cna_subset = cna_subset[complete.cases(cna_subset),,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients_
  overlap = intersect(rownames(clin_sig), rownames(cna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  cna_subset = cna_subset[rownames(cna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  cna_subset = cna_subset[order(rownames(cna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(cna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  plots = list()
  stats = list() # Holds the mean and variance of the counts for each subtype.
  results = list() # Holds the negative binomial regression model of the counts for each subtype. 
  for(subtype in names(subtypes)) {
    # Subset both cna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    cna_subset_subtype = cna_subset[rownames(cna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
    ## Divide up the samples by copy number of the genes in the signature.
    cna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    # for(k in 1:nrow(cna_subset)) {
    #      if(cna_subset[k,1] < cna_subset[k,2]) { cna_subset_group[k] = paste0("Low ubiquitination")} 
    #      else if(cna_subset[k,1] > cna_subset[k,2]) { cna_subset_group[k] = paste0("High ubiquitination")}
    #      else { cna_subset_group[k] = NA}
    #      
    #    }
    for(k in 1:nrow(cna_subset_subtype)) {
      if(cna_subset_subtype[k,1] < 0) { cna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(cna_subset_subtype[k,1] > 0) { cna_subset_group[k] = paste0("High ", sig_gene)} 
      else { cna_subset_group[k] = paste0("Neutral ", sig_gene) }
    }
    cna_subset_subtype$Group = cna_subset_group
    # Remove NAs.
    cna_subset_subtype = cna_subset_subtype[complete.cases(cna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(cna_subset_subtype)),]
    identical(rownames(cna_subset_subtype), rownames(clin_sig_subtype))
    
    # Create a table holding, for each sample, 1) gene CN group, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
    other_clin_features = data.frame(UbSig=cna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
    
    # Re-order the levels of "UbSig": High - Neutral - Low.
    other_clin_features$UbSig = factor(other_clin_features$UbSig, levels = c(paste0("Low ", sig_gene), paste0("Neutral ", sig_gene), paste0("High ", sig_gene)))
    
    # Get the mean and variance of the counts.
    stats_table = data.frame(
      Subtype = subtype,
      "Mean number of positive lymph nodes" = round(mean(other_clin_features$NumLymphNodes, na.rm = T), 2),
      "Variance" = round(var(other_clin_features$NumLymphNodes, na.rm = T), 2)
      )
    colnames(stats_table) = base::gsub("\\.", " ", colnames(stats_table))
    stats[[subtype]] = stats_table
    
    # Build a negative binomial regression model.
    # https://data.princeton.edu/wws509/r/overdispersion
    # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
    # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
    # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
    mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
    p_vals = coef(summary(mnb))[,4]
    
    # Graph.
    # http://www.cookbook-r.com/Graphs/Plotting_distributions_(ggplot2)/
    lymph_nodes = other_clin_features[,c("UbSig", "NumLymphNodes")]
    nodes_plot = ggplot(lymph_nodes, aes(x=NumLymphNodes, fill=UbSig)) + 
      geom_histogram(position="identity", stat = "count") + 
      facet_grid(UbSig ~ ., margins = TRUE, scales = "free") + 
      #geom_boxplot() + 
      labs(fill = "Group") +
      #annotate("text", x = 4, y = 25, label = "italic(R) ^ 2 == 0.75", parse = TRUE) + # Annotate with p-value.
      scale_fill_manual(values=c("#0073C2FF", "#868686FF", "#EFC000FF", "#CD534CFF")) +
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size),
            strip.text.y = element_blank()) +
      labs(x=" ", y=" ", title=paste0(subtype))
    
    # Add plots to plots list.
    plots[["Nodes"]][[subtype]] = nodes_plot
    # Add statistical test results to list. 
    results[["Nodes"]][["Test"]][[subtype]] = summary(mnb)
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      Count = sum(!is.na(other_clin_features$NumLymphNodes)),
      stringsAsFactors = F
      ))
    }
  saveRDS(plots, file = paste0("FigureS9_", sig_gene, "_plot.rds"))
  saveRDS(stats, file = paste0("FigureS9_", sig_gene, "_descriptive_statistics.rds"))
  saveRDS(results, file = paste0("FigureS9_", sig_gene, "_inferential_statistics.rds"))

  annotated_fig_list = list()
  for(variable in c("Nodes")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[variable]], 
                               nrow = 2, 
                               ncol = 2, 
                               #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                               #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                               font.label = list(size = 20),
                               common.legend = TRUE, 
                               legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                       top = text_grob(paste("Association of", sig_gene, "copy number with", str_to_lower(full_variable)), size = overall_title_size),
                                       left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                       bottom = text_grob(paste("Number of positive lymph nodes"), size = common_axis_text_x_size),
                                       fig.lab = letters[j], 
                                       fig.lab.size = figure_label_size,
                                       fig.lab.face = figure_label_face
                                       )
    annotated_fig_list[[variable]] = figure_annotated
  
  }
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S9", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Nodes)
  dev.off()
  
}

```

### Figure S10 - Clinical association of expression of signature genes individually (cf. Figure 7)
```{r}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

## Fetch and clean the data.
# Create a CGDS object.
mycgds = CGDS("http://www.cbioportal.org/")

upper_quantile_sig = 0.50
lower_quantile_sig = 0.50
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  
  # Get the expression data.
  mrna_subset = getProfileData(mycgds, sig_gene, "brca_metabric_mrna", "brca_metabric_mrna")
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_mrna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  # Throw out all the samples that don't have mrna data.
  mrna_subset = mrna_subset[complete.cases(mrna_subset),,drop=FALSE]
  # Make sure both the clinical and expression data have the same patients_
  overlap = intersect(rownames(clin_sig), rownames(mrna_subset))
  clin_sig = clin_sig[which(rownames(clin_sig) %in% overlap),,drop=FALSE]
  mrna_subset = mrna_subset[rownames(mrna_subset) %in% overlap,,drop=FALSE]
  # Order rownames so they're both in the same order.
  mrna_subset = mrna_subset[order(rownames(mrna_subset)),,drop=FALSE]
  clin_sig = clin_sig[order(rownames(clin_sig)),]
  identical(rownames(mrna_subset), rownames(clin_sig))
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  plots = list()
  stats = list() # Holds the mean and variance of the counts for each subtype.
  results = list() # Holds the negative binomial regression model of the counts for each subtype. 
  for(subtype in names(subtypes)) {
    # Subset both mrna_subset and clin_sig to include only the samples in the current subtype. 
    sample_IDs = subtypes[[subtype]]
    mrna_subset_subtype = mrna_subset[rownames(mrna_subset) %in% sample_IDs,,drop=FALSE]
    clin_sig_subtype = clin_sig[rownames(clin_sig) %in% sample_IDs,]
    
     ## Divide up the samples by copy number of the genes in the signature.
    mrna_subset_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample.
    for(k in 1:nrow(mrna_subset_subtype)) {
      if(mrna_subset_subtype[k,1] <= quantile(mrna_subset_subtype[,1], lower_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("Low ", sig_gene)} 
      else if(mrna_subset_subtype[k,1] > quantile(mrna_subset_subtype[,1], upper_quantile_sig, na.rm=TRUE)) { mrna_subset_group[k] = paste0("High ", sig_gene)} 
      else { mrna_subset_group[k] = NA }
    }
    mrna_subset_subtype$Group = mrna_subset_group
    # Remove NAs.
    mrna_subset_subtype = mrna_subset_subtype[complete.cases(mrna_subset_subtype),]
    clin_sig_subtype = clin_sig_subtype[which(rownames(clin_sig_subtype) %in% rownames(mrna_subset_subtype)),]
    identical(rownames(mrna_subset_subtype), rownames(clin_sig_subtype))
    
    # Re-order the levels of "Group": Low - High.
    mrna_subset_subtype$Group = factor(mrna_subset_subtype$Group, levels = c(paste0('Low ', sig_gene), paste0('High ', sig_gene)))
    
    # Create a table holding, for each sample, 1) gene CN group, 2) stage, 3) grade, 4) number of lymph nodes, and 5) size.
    other_clin_features = data.frame(UbSig=mrna_subset_group, Stage=clin_sig_subtype$TUMOR_STAGE, Grade=clin_sig_subtype$GRADE, NumLymphNodes=clin_sig_subtype$LYMPH_NODES_EXAMINED_POSITIVE, Size=clin_sig_subtype$TUMOR_SIZE)
    
    # Re-order the levels of "UbSig": High - Low.
    other_clin_features$UbSig = factor(other_clin_features$UbSig, levels = c(paste0("Low ", sig_gene), paste0("High ", sig_gene)))
    
    # Get the mean and variance of the counts.
    stats_table = data.frame(
      Subtype = subtype,
      "Mean number of positive lymph nodes" = round(mean(other_clin_features$NumLymphNodes, na.rm = T), 2),
      "Variance" = round(var(other_clin_features$NumLymphNodes, na.rm = T), 2)
      )
    colnames(stats_table) = base::gsub("\\.", " ", colnames(stats_table))
    stats[[subtype]] = stats_table
    
    # Build a negative binomial regression model.
    # https://data.princeton.edu/wws509/r/overdispersion
    # https://stats.idre.ucla.edu/stata/output/negative-binomial-regression/ (See section "Female.")
    # https://stats.stackexchange.com/questions/48448/how-can-i-interpret-coefficients-of-categorical-predictors-in-the-negative-binom
    # https://stats.idre.ucla.edu/r/dae/negative-binomial-regression/
    mnb = glm.nb(NumLymphNodes ~ UbSig, data = other_clin_features)
    p_vals = coef(summary(mnb))[,4]
    
    # Graph.
    # http://www.cookbook-r.com/Graphs/Plotting_distributions_(ggplot2)/
    lymph_nodes = other_clin_features[,c("UbSig", "NumLymphNodes")]
    nodes_plot = ggplot(lymph_nodes, aes(x=NumLymphNodes, fill=UbSig)) + 
      geom_histogram(position="identity", stat = "count") + 
      facet_grid(UbSig ~ ., margins = TRUE, scales = "free") + 
      #geom_boxplot() + 
      labs(fill = "Group") +
      #annotate("text", x = 4, y = 25, label = "italic(R) ^ 2 == 0.75", parse = TRUE) + # Annotate with p-value.
      scale_fill_manual(values=c("#0073C2FF", "#EFC000FF", "#CD534CFF")) +
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size),
            strip.text.y = element_blank()) +
      labs(x=" ", y=" ", title=paste0(subtype))
    
    # Add plots to plots list.
    plots[["Nodes"]][[subtype]] = nodes_plot
    # Add statistical test results to list. 
    results[["Nodes"]][["Test"]][[subtype]] = summary(mnb)
    
    # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = sig_gene,
      Subtype = subtype,
      Count = sum(!is.na(other_clin_features$NumLymphNodes)),
      stringsAsFactors = F
      ))
    }
  saveRDS(plots, file = paste0("FigureS10_", sig_gene, "_plot.rds"))
  saveRDS(stats, file = paste0("FigureS10_", sig_gene, "_descriptive_statistics.rds"))
  saveRDS(results, file = paste0("FigureS10_", sig_gene, "_inferential_statistics.rds"))

  annotated_fig_list = list()
  for(variable in c("Nodes")) {
    full_variable = ifelse(variable=="Nodes", "Number of Positive Lymph Nodes", variable)
    
    # Arrange plots.
    figure = ggpubr::ggarrange(plotlist = plots[[variable]], 
                               nrow = 2, 
                               ncol = 2, 
                               #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                               #hjust = c(-0.5, -1, -0.5, -1, -1.5),
                               font.label = list(size = 20),
                               common.legend = TRUE, 
                               legend = "right")
    #n = length(plots[[variable]])
    #n_col = floor(sqrt(n))
    #do.call("grid.arrange", c(plots[[variable]], ncol = n_col))
    # Annotate figure by adding a common label.
    # https://github.com/kassambara/ggpubr/issues/78
    common_y_axis_label = ifelse(variable=="Nodes", "Patient count", "Proportion of patients")
    figure_annotated = annotate_figure(figure,
                                       top = text_grob(paste("Association of", sig_gene, "expression with", str_to_lower(full_variable)), size = overall_title_size),
                                       left = text_grob(paste(common_y_axis_label), rot = 90, size = common_axis_text_y_size),
                                       bottom = text_grob(paste("Number of positive lymph nodes"), size = common_axis_text_x_size),
                                       fig.lab = letters[j], 
                                       fig.lab.size = figure_label_size,
                                       fig.lab.face = figure_label_face
                                       )
    annotated_fig_list[[variable]] = figure_annotated
    
  }
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S10", letters[j], ".eps"), width = 20, height = 20)
  print(annotated_fig_list$Nodes)
  dev.off()
  
}

```

### Figure S11 - Association between copy number and expression
```{r}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

plots = list()
dat_tables = list()
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  plots[[sig_gene]] = list()
  dat_tables[[sig_gene]] = list()
  
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_mrna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  # For each subtype ... 
  for(subtype in names(subtypes)) {
    # Set up the data frame that will hold the data for all the cancers, for subtype "subtype". 
    dat = data.frame(Genes=character(), Expression=numeric(), Group=character())
    # Get the sample IDs for this subtype.
    sample_IDs = base::gsub("\\-", "\\.", subtypes[[subtype]])
    
    # Get all the samples that have expression.
    mrna = getProfileData(mycgds, sig_gene, paste0(tolower(cancer), "_metabric_mrna"), paste0(tolower(cancer), "_metabric_mrna"))[,1,drop=FALSE]
    # Remove all samples that do not have expression data.
    mrna = mrna[complete.cases(mrna),,drop=FALSE]
    # if(nrow(mrna)==0) next # If no mrnaein data, skip to the next gene.
    
    # Get the CNA data of the predictor genes.
    cna = getProfileData(mycgds, sig_gene, paste0(tolower(cancer), "_metabric_cna"), paste0(tolower(cancer), "_metabric_cna"))
    # Filter out incomplete cases. 
    # Remove incomplete cases (NAs or NaNs) and log-transform.
    cna = cna[complete.cases(cna),,drop=FALSE]
    if(nrow(cna)==0) next # If no cna data, skip to the next gene. 
    
    # Keep only the samples that have both gene and mrna levels. 
    good_samples = intersect(intersect(rownames(cna), rownames(mrna)), sample_IDs)
    mrna = mrna[which(rownames(mrna) %in% good_samples),,drop=FALSE]
    cna = cna[which(rownames(cna) %in% good_samples),,drop=FALSE]
    # if(nrow(mrna)==0 | nrow(cna)==0) next 
    
    # Make sure the samples are in the same order in both the mrna and gene copy number vectors.
    mrna = mrna[order(rownames(mrna)),,drop=FALSE]
    cna = cna[order(rownames(cna)),,drop=FALSE]
    # identical(rownames(mrna), rownames(cna))
    
    # Divide samples by copy number of the predictor genes.
    # Create the signature matrix.
    group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
    for(k in 1:nrow(cna)) {
      if(cna[k,1] == -2) { group[k] = paste0("Deep deletion")} 
      else if(cna[k,1] == -1) { group[k] = paste0("Shallow deletion")}
      else if(cna[k,1] == 0 ) {group[k] = paste0("Diploid")}
      else if(cna[k,1] == 1) {group[k] = paste0("Gain")}
      else { group[k] = paste0("Amplification")}
    }
    # Bind the grouping to the matrix.
    cna$Group = as.factor(group)
    # Keep only complete cases.
    cna = cna[complete.cases(cna),]
    mrna = mrna[rownames(mrna) %in% rownames(cna),,drop=FALSE]
    if(identical(rownames(cna), rownames(mrna))) {
      mrna$Group = cna$Group
    }
    
    dat = rbind(dat, data.frame(Genes=paste0(sig_gene), Expression=mrna[,1], Group=mrna$Group))
    # Re-order the levels of "Group": Low - High.
   dat$Group = factor(dat$Group, levels = c("Deep deletion", "Shallow deletion", "Diploid", "Gain", "Amplification"))
   
   # Save the counts.
   counts_table_final = rbind(counts_table_final, data.frame(
     Gene = gene,
     Subtype = subtype,
     Count = nrow(dat),
     stringsAsFactors = F
     ))
      
    # Save dat to the list.
    dat_tables[[sig_gene]][[subtype]] = dat
    write.table(dat, file = paste0("FigureS11_", sig_gene, "_", subtype, ".csv"), sep = ",", row.names = F, col.names = F)
    
    # Use ggplot2 to create a dotplot. 
    plots[[sig_gene]][[subtype]] = ggplot(data=dat, aes(x=Genes, y=Expression, fill=Group, group=interaction(Group, Genes))) + # Each separate boxplot will be the "intersection" (i.e. interaction) of a gene (or gene signature) x group.
      geom_boxplot(lwd = boxplot_lwd) + # geom_violin() # https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
   scale_fill_manual(breaks = c('Deep deletion', 'Shallow deletion', 'Diploid', 'Gain', 'Amplification'),
                        values=pal_jco()(5)) + #scale_fill_jco() +
      labs(title=subtype, x=" ", y=" ") #+ 
    #stat_compare_means(label =  "p.signif", 
    #                   label.x = 1.5, 
    #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
    #                                      symbols = c("****", "***", "**", "*", " ")), 
    #                   size = p_value_size )  
  }
  
  # Arrange plots.
  figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]], 
                             nrow = 2, 
                             ncol = 2, 
                             #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                             hjust = c(-0.5, -1, -0.5, -1, -1.5),
                             font.label = list(size = 20),
                             common.legend = TRUE, 
                             legend = "right")
  # Annotate figure by adding a common label.
  # https://github.com/kassambara/ggpubr/issues/78
  figure_S11_j = annotate_figure(figure,
                                   top = text_grob(paste(sig_gene, "expression between copy-number groups"), size = overall_title_size),
                                   left = text_grob(paste(sig_gene, "expression"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Gene(s)"), size = common_axis_text_x_size),
                                   fig.lab = letters[j],
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
                               )
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S11", letters[j], ".eps"), width = 20, height = 20)
  print(figure_S11_j)
  dev.off()
  
}

# Statistics. 
stats = list()
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  stats[[sig_gene]] = list()
  # http://www.sthda.com/english/wiki/one-way-anova-test-in-r
  for(subtype in names(subtypes)) {
    dat = dat_tables[[sig_gene]][[subtype]]
    dat_sub = dat %>% dplyr::filter(Genes==sig_gene) %>% droplevels()
    
    # Compute summary statistics.
    summary_stats = dat_sub %>% group_by(Group) %>%
      summarise(
        Count = n(),
        Mean = mean(Expression, na.rm = TRUE) %>% signif(digits = 2),
        SD = sd(Expression, na.rm = TRUE) %>% signif(digits = 2)
        )
    # Save to stats list.
    stats[[sig_gene]][[subtype]][["Summary"]] = summary_stats
    
    # Perform ANOVA.
    res_aov = aov(Expression ~ Group, data = dat_sub)
    # Summary of the analysis.
    aov_stats = summary(res_aov)
    # Save to stats list.
    stats[[sig_gene]][[subtype]][["Inferential"]] = aov_stats
  }
  
  saveRDS(stats, file = paste0("FigureS11", letters[j], "_statistics.rds"))
}

```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = " "}
figure_S15_j
```

### Figure S12 (cf. Figure 2)
```{r echo = FALSE, eval = TRUE, include=FALSE}
counts_table_final = data.frame(
  Gene = character(),
  Subtype = character(),
  Count = integer(),
  stringsAsFactors = F
)

# Divide the patients into the four molecular subtypes as listed here: https://ww5.komen.org/BreastCancer/SubtypesofBreastCancer.html. 
# On using .$ to extract column as vector: https://stackoverflow.com/a/27160317 
clin_dat = read.delim("brca_tcga_clinical_data.tsv", stringsAsFactors = F)
subtypes = list()
subtypes[["All"]] = clin_dat %>% .$Sample.ID
subtypes[["Luminal"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive") %>% .$Sample.ID
subtypes[["HER2"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Positive") %>% .$Sample.ID
#subtypes[["Non-TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Positive" | PR.status.by.ihc=="Positive" | IHC.HER2=="Positive") %>% .$Sample.ID
subtypes[["TNBC"]] = clin_dat %>% filter(ER.Status.By.IHC=="Negative" & PR.status.by.ihc=="Negative" & IHC.HER2=="Negative") %>% .$Sample.ID

# CNA
# For each subtype ... 
plots = list()
dat_tables = list()
downstream_gene = "CDKN1A"
prot_name = dict[[downstream_gene]]
for(subtype in names(subtypes)) {
  # Set up the data frame that will hold the data for all the cancers, for subtype "subtype". 
  dat = data.frame(Genes=character(), ProteinLevels=numeric(), Group=character())
  # Get the sample IDs for this subtype.
  sample_IDs = base::gsub("\\-", "\\.", subtypes[[subtype]])
  
  # For each gene in dub_genes, 1) get for all the samples in the subtype a) the copy numbers of the dub gene, b) the copy numbers of the ub gene, and c) the protein levels of p27; 2) split the samples into two groups by copy-number status (copy number of dub gene > copy number of ub gene and v.v.); 3) compare the p27 levels between the two groups. 
  for(dub_gene in dub_genes) {
    # Get all the samples that have target protein expression.
    #prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_rppa"))[,1,drop=FALSE]
    prot = getProfileData(mycgds, downstream_gene, paste0(tolower(cancer), "_tcga_rppa"), paste0(tolower(cancer), "_tcga_all"))[,1,drop=FALSE]
    # Remove all samples that do not have expression data.
    prot = prot[complete.cases(prot),,drop=FALSE]
    # if(nrow(prot)==0) next # If no protein data, skip to the next gene.
    
    # Get the CNA data of the predictor genes.
    cna = getProfileData(mycgds, c(dub_gene, ub_genes[1]), paste0(tolower(cancer), "_tcga_gistic"), paste0(tolower(cancer), "_tcga_cna"))
    # Filter out incomplete cases. 
    # Remove incomplete cases (NAs or NaNs) and log-transform.
    cna = cna[complete.cases(cna),,drop=FALSE]
    if(nrow(cna)==0) next # If no cna data, skip to the next gene. 
    
    # Keep only the samples that have both gene and protein levels. 
    good_samples = intersect(intersect(rownames(cna), rownames(prot)), sample_IDs)
    prot = prot[which(rownames(prot) %in% good_samples),,drop=FALSE]
    cna = cna[which(rownames(cna) %in% good_samples),,drop=FALSE]
    # if(nrow(prot)==0 | nrow(cna)==0) next 
    
    # Make sure the samples are in the same order in both the protein and gene copy number vectors.
    prot = prot[order(rownames(prot)),,drop=FALSE]
    cna = cna[order(rownames(cna)),,drop=FALSE]
    # identical(rownames(prot), rownames(cna))
    # Make sure the column names of the predictor genes are in alphabetical order.
    cna = cna[,order(colnames(cna), decreasing = FALSE),drop=FALSE]
    
    # Divide samples by copy number of the predictor genes.
    # Create the signature matrix.
    group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
    for(k in 1:nrow(cna)) {
      if(cna[k,1] < cna[k,2]) { group[k] = paste0("Low ubiquitination")} 
      else if(cna[k,1] > cna[k,2]) { group[k] = paste0("High ubiquitination")}
      else { group[k] = paste0("Intermediate ubiquitination")}
      
    }
    # Bind the grouping to the matrix.
    cna$Group = as.factor(group)
    # Keep only complete cases.
    cna = cna[complete.cases(cna),]
    prot = prot[rownames(prot) %in% rownames(cna),,drop=FALSE]
    if(identical(rownames(cna), rownames(prot))) {
      prot$Group = cna$Group
    }
    
    dat = rbind(dat, data.frame(Genes=paste0(dub_gene, ", ", ub_genes[1]), ProteinLevels=prot[,1], Group=prot$Group))
  
  }
  
  # Save the counts.
    counts_table_final = rbind(counts_table_final, data.frame(
      Gene = dub_gene,
      Subtype = subtype,
      Count = nrow(dat),
      stringsAsFactors = F
      ))
  
  # Save dat to the list.
  dat_tables[[subtype]] = dat
  write.table(dat, file = paste0("FigureS12_CNA_", subtype, ".csv"), sep = ",", row.names = F, col.names = F)
  
  # Use ggplot2 to create a dotplot. 
   plots[[subtype]] = ggplot(data=dat, aes(x=Genes, y=ProteinLevels, fill=Group, group=interaction(Group, Genes))) + # Each separate boxplot will be the "intersection" (i.e. interaction) of a gene (or gene signature) x group.
    geom_boxplot(lwd = boxplot_lwd) + # geom_violin() # https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
    theme_classic() + 
    theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
          axis.line = element_line(size = axis_line_size),
          axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
          axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
          axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
          axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
          legend.title = element_text(size = legend_title_size),
          legend.text = element_text(size = legend_text_size)) +
    scale_fill_manual(values=c("#0073C2FF", "#868686FF", "#EFC000FF")) + #scale_fill_jco() +
    labs(title=subtype, x=" ", y=" ") #+ 
    #stat_compare_means(label =  "p.signif", 
    #                   label.x = 1.5, 
    #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
    #                                      symbols = c("****", "***", "**", "*", " ")), 
    #                   size = p_value_size )  
}

# Arrange plots.
figure = ggpubr::ggarrange(plotlist = plots, 
                   nrow = 2, 
                   ncol = 2, 
                   #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                   hjust = c(-0.5, -1, -0.5, -1, -1.5),
                   font.label = list(size = 20),
                   common.legend = TRUE, 
                   legend = "right")
# Annotate figure by adding a common label.
# https://github.com/kassambara/ggpubr/issues/78
figure_2 = annotate_figure(figure,
                                   top = text_grob(paste(prot_name, "protein levels between SKP2 ubiquitination signature groups"), size = overall_title_size),
                                   left = text_grob(paste(prot_name, "protein levels"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Gene(s)"), size = common_axis_text_x_size),
                                   fig.lab = " ", 
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
)

# Save to EPS.
setEPS()
postscript("Figure_S12.eps", width = 20, height = 20)
figure_2
dev.off()

# Statistics.
# http://www.sthda.com/english/wiki/one-way-anova-test-in-r
stats = list()
for(subtype in names(subtypes)) {
  dat = dat_tables[[subtype]]
  
  for(gene_set in unique(dat$Genes)) {
    dat_sub = dat %>% dplyr::filter(Genes==gene_set) %>% droplevels()
    
    # Compute summary statistics.
    summary_stats = dat_sub %>% group_by(Group) %>%
      summarise(
        Count = n(),
        Mean = mean(ProteinLevels, na.rm = TRUE) %>% signif(digits = 2),
        SD = sd(ProteinLevels, na.rm = TRUE) %>% signif(digits = 2)
        )
    # Save to stats list.
    stats[[subtype]][[gene_set]][["Summary"]] = summary_stats
    
    # Perform ANOVA.
    res_aov = aov(ProteinLevels ~ Group, data = dat_sub)
    # Summary of the analysis.
    aov_stats = summary(res_aov)
    # Save to stats list.
    stats[[subtype]][[gene_set]][["Inferential"]] = aov_stats
    
  }
}
saveRDS(stats, file = paste0("FigureS12_statistics.rds"))

# show_col(pal_jco()(4))
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = ""}
plotlist = list(figure_2)
grid.arrange(grobs = lapply(plotlist, "+", margin), nrow = 1)
```

### Figure S13 - Association between copy number and expression for SKP2 (not included in Supplemental)
```{r}
## Set the genes for the signature.
genes_sig = c("SKP2")

plots = list()
dat_tables = list()
stats = list()
for(j in 1:length(genes_sig)) {
  sig_gene = genes_sig[j]
  plots[[sig_gene]] = list()
  dat_tables[[sig_gene]] = list()
  stats[[sig_gene]] = list()
  
  # Get the clinical data.
  clin_sig = getClinicalData(mycgds, "brca_metabric_mrna")
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  
  clin_sig$SAMPLE_ID = rownames(clin_sig)
  subtypes = list()
  subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID
  subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  #subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
  subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
  
  # For each subtype ... 
  for(subtype in names(subtypes)) {
    # Set up the data frame that will hold the data for all the cancers, for subtype "subtype". 
    dat = data.frame(Genes=character(), Expression=numeric(), Group=character())
    # Get the sample IDs for this subtype.
    sample_IDs = base::gsub("\\-", "\\.", subtypes[[subtype]])
    
    # Get all the samples that have expression.
    mrna = getProfileData(mycgds, sig_gene, paste0(tolower(cancer), "_metabric_mrna"), paste0(tolower(cancer), "_metabric_mrna"))[,1,drop=FALSE]
    # Remove all samples that do not have expression data.
    mrna = mrna[complete.cases(mrna),,drop=FALSE]
    # if(nrow(mrna)==0) next # If no mrnaein data, skip to the next gene.
    
    # Get the CNA data of the predictor genes.
    cna = getProfileData(mycgds, sig_gene, paste0(tolower(cancer), "_metabric_cna"), paste0(tolower(cancer), "_metabric_cna"))
    # Filter out incomplete cases. 
    # Remove incomplete cases (NAs or NaNs) and log-transform.
    cna = cna[complete.cases(cna),,drop=FALSE]
    if(nrow(cna)==0) next # If no cna data, skip to the next gene. 
    
    # Keep only the samples that have both gene and mrna levels. 
    good_samples = intersect(intersect(rownames(cna), rownames(mrna)), sample_IDs)
    mrna = mrna[which(rownames(mrna) %in% good_samples),,drop=FALSE]
    cna = cna[which(rownames(cna) %in% good_samples),,drop=FALSE]
    # if(nrow(mrna)==0 | nrow(cna)==0) next 
    
    # Make sure the samples are in the same order in both the mrna and gene copy number vectors.
    mrna = mrna[order(rownames(mrna)),,drop=FALSE]
    cna = cna[order(rownames(cna)),,drop=FALSE]
    # identical(rownames(mrna), rownames(cna))
    
    # Divide samples by copy number of the predictor genes.
    # Create the signature matrix.
    group = c() # The vector that holds the grouping (high vs. low copy number) for each sample.
    for(k in 1:nrow(cna)) {
      if(cna[k,1] == -2) { group[k] = paste0("Deep deletion")} 
      else if(cna[k,1] == -1) { group[k] = paste0("Shallow deletion")}
      else if(cna[k,1] == 0 ) {group[k] = paste0("Diploid")}
      else if(cna[k,1] == 1) {group[k] = paste0("Gain")}
      else { group[k] = paste0("Amplification")}
    }
    # Bind the grouping to the matrix.
    cna$Group = as.factor(group)
    # Keep only complete cases.
    cna = cna[complete.cases(cna),]
    mrna = mrna[rownames(mrna) %in% rownames(cna),,drop=FALSE]
    if(identical(rownames(cna), rownames(mrna))) {
      mrna$Group = cna$Group
    }
    
    dat = rbind(dat, data.frame(Genes=paste0(sig_gene), Expression=mrna[,1], Group=mrna$Group))
    # Re-order the levels of "Group": Low - High.
   dat$Group = factor(dat$Group, levels = c("Deep deletion", "Shallow deletion", "Diploid", "Gain", "Amplification"))
    
    # Save dat to the list.
    dat_tables[[sig_gene]][[subtype]] = dat
    write.table(dat, file = paste0("FigureS13_", sig_gene, "_", subtype, ".csv"), sep = ",", row.names = F, col.names = F)
    
    # Use ggplot2 to create a dotplot. 
    plots[[sig_gene]][[subtype]] = ggplot(data=dat, aes(x=Genes, y=Expression, fill=Group, group=interaction(Group, Genes))) + # Each separate boxplot will be the "intersection" (i.e. interaction) of a gene (or gene signature) x group.
      geom_boxplot(lwd = boxplot_lwd) + # geom_violin() # https://stackoverflow.com/questions/23433776/change-thickness-of-the-whole-line-geom-boxplot
      theme_classic() + 
      theme(plot.title = element_text(size = title_size, face = title_face, margin = title_margins),
            axis.line = element_line(size = axis_line_size),
            axis.text.x = element_text(color = axis_text_x_color, size = axis_text_x_size, hjust = .5, vjust = .5, face = axis_text_x_face),
            axis.title.x = element_text(color = axis_title_x_color, size = axis_title_x_size, hjust = .5, vjust = .5, face = axis_title_x_face),
            axis.text.y = element_text(color = axis_text_y_color, size = axis_text_y_size, hjust = .5, vjust = .5, face = axis_text_y_face),
            axis.title.y = element_text(color = axis_title_y_color, size = axis_title_y_size, hjust = .5, vjust = .5, face = axis_title_y_face),
            legend.title = element_text(size = legend_title_size),
            legend.text = element_text(size = legend_text_size)) +
   scale_fill_manual(breaks = c('Deep deletion', 'Shallow deletion', 'Diploid', 'Gain', 'Amplification'),
                        values=pal_jco()(5)) + #scale_fill_jco() +
      labs(title=subtype, x=" ", y=" ") #+ 
    #stat_compare_means(label =  "p.signif", 
    #                   label.x = 1.5, 
    #                   symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
    #                                      symbols = c("****", "***", "**", "*", " ")), 
    #                   size = p_value_size )  
  }
  
  # Arrange plots.
  figure = ggpubr::ggarrange(plotlist = plots[[sig_gene]], 
                             nrow = 2, 
                             ncol = 2, 
                             #labels = c("Luminal", "HER2", "Non-TNBC", "TNBC", "All"),
                             hjust = c(-0.5, -1, -0.5, -1, -1.5),
                             font.label = list(size = 20),
                             common.legend = TRUE, 
                             legend = "right")
  # Annotate figure by adding a common label.
  # https://github.com/kassambara/ggpubr/issues/78
  figure_S13_j = annotate_figure(figure,
                                   top = text_grob(paste(sig_gene, "expression between copy-number groups"), size = overall_title_size),
                                   left = text_grob(paste(sig_gene, "expression"), rot = 90, size = common_axis_text_y_size),
                                   bottom = text_grob(paste("Gene(s)"), size = common_axis_text_x_size),
                                   fig.lab = "", #letters[j],
                                   fig.lab.size = figure_label_size,
                                   fig.lab.face = figure_label_face
                               )
  
  # Save to EPS.
  setEPS()
  postscript(paste0("Figure_S13", letters[j], ".eps"), width = 20, height = 20)
  print(figure_S13_j)
  dev.off()
  
  # Statistics.
  # http://www.sthda.com/english/wiki/one-way-anova-test-in-r
  for(subtype in names(subtypes)) {
    dat = dat_tables[[sig_gene]][[subtype]]
    dat_sub = dat %>% dplyr::filter(Genes==sig_gene) %>% droplevels()
    
    # Compute summary statistics.
    summary_stats = dat_sub %>% group_by(Group) %>%
      summarise(
        Count = n(),
        Mean = mean(Expression, na.rm = TRUE) %>% signif(digits = 2),
        SD = sd(Expression, na.rm = TRUE) %>% signif(digits = 2)
        )
    # Save to stats list.
    stats[[sig_gene]][[subtype]][["Summary"]] = summary_stats
    
    # Perform ANOVA.
    res_aov = aov(Expression ~ Group, data = dat_sub)
    # Summary of the analysis.
    aov_stats = summary(res_aov)
    # Save to stats list.
    stats[[sig_gene]][[subtype]][["Inferential"]] = aov_stats
  }
  
  saveRDS(stats, file = paste0("FigureS13", letters[j], "_statistics.rds"))
  
}
```

```{r echo = FALSE, eval = TRUE, fig.width = 20, fig.height = 20, fig.margin = TRUE, fig.cap = " "}
figure_S17_j
```

### Table S1 (Figure 2)
```{r echo = FALSE, eval = TRUE, fig.cap=""}
# Load the statistics.
stats = readRDS(paste0("Figure2_statistics.rds"))

ft_list = list()
for(subtype in names(stats)) {
  for(gene_list in names(stats[[subtype]])) {
   descriptive_stats = stats[[subtype]][[gene_list]]$Summary %>% as.data.frame
   ft = flextable(descriptive_stats)
   ft = add_header_lines(ft, values = gene_list)
   ft = add_header_lines(ft, values = subtype, top = TRUE)
   ft_list[[subtype]][[gene_list]] = autofit(ft) 
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$All$`USP13, FZR1`
ft_list$All$`USP10, FZR1`
ft_list$All$`USP14, FZR1`
ft_list$Luminal$`USP13, FZR1`
ft_list$Luminal$`USP10, FZR1`
ft_list$Luminal$`USP14, FZR1`
ft_list$HER2$`USP13, FZR1`
ft_list$HER2$`USP10, FZR1`
ft_list$HER2$`USP14, FZR1`
ft_list$TNBC$`USP13, FZR1`
ft_list$TNBC$`USP10, FZR1`
ft_list$TNBC$`USP14, FZR1`
```

### Table S2 (Figure 4a)
```{r echo = FALSE, eval=TRUE, include=FALSE}
stats = readRDS("Figure4a_statistics.rds")

ft_list = list()
for(subtype in names(stats)) {
  tmp=capture.output(print(stats[[subtype]], print.n="records", show.rmean=FALSE))
  tmp[4] = "Group                                    n events median 0.95LCL 0.95UCL"
  table = read.table(textConnection(tmp), skip=3, header=TRUE)
  table[,1] = table[,1] %>% regexPipes::gsub("Low", "Low ") %>% regexPipes::gsub("Neutral", "Neutral ") %>% regexPipes::gsub("High", "High ")
  colnames(table) = colnames(table) %>% regexPipes::gsub("X0\\.95", "0\\.95\\% ") %>% regexPipes::gsub("events", "Events") %>% regexPipes::gsub("median", "Median") 
  
  ft = flextable(table)
  ft = add_header_lines(ft, values = subtype)
  ft = add_header_lines(ft, values = sig_gene)
  ft_list[[subtype]] = autofit(ft)
    
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$All
ft_list$Luminal
ft_list$HER2
ft_list$TNBC
```

### Table S3 (Figure 4b)
```{r echo = FALSE, eval=TRUE, include=FALSE}
stats = readRDS("Figure4b_statistics.rds")

tmp=capture.output(print(stats, print.n="records", show.rmean=FALSE))
tmp[4] = "Group                                    n events median 0.95LCL 0.95UCL"
table = read.table(textConnection(tmp), skip=3, header=TRUE)
table[,1] = table[,1] %>% regexPipes::gsub("Low", "Low ") %>% regexPipes::gsub("Neutral", "Neutral ") %>% regexPipes::gsub("High", "High ")
colnames(table) = colnames(table) %>% regexPipes::gsub("X0\\.95", "0\\.95\\% ") %>% regexPipes::gsub("events", "Events") %>% regexPipes::gsub("median", "Median") 
  
ft = flextable(table)
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft
```

### Table S4 (Figure S2 - Association of CNA of single genes with p27 protein levels)
```{r echo = FALSE, eval = TRUE, fig.cap=""}
# Load the statistics.
stats = readRDS(paste0("FigureS2_statistics.rds"))

apply(coef(results$Nodes$Test[[subtype]]), c(1,2), signif, digits = 2)

ft_list = list()
for(gene in names(stats)) {
  for(subtype in names(stats[[gene]])) {
    descriptive_stats_numeric = stats[[gene]][[subtype]]$Summary %>% .[,3:4] %>% apply(c(1,2), signif, digits = 2) %>% as.data.frame
    descriptive_stats_non_numeric = stats[[gene]][[subtype]]$Summary %>% .[,1:2] %>% as.data.frame
   descriptive_stats = cbind(descriptive_stats_non_numeric, descriptive_stats_numeric)
   ft = flextable(descriptive_stats)
   ft = add_header_lines(ft, values = subtype)
   ft = add_header_lines(ft, values = gene, top = TRUE)
   ft_list[[gene]][[subtype]] = autofit(ft) 
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC
```

### Table S5 (Figure S3 - Survival analysis - CNA groups)
```{r echo = FALSE, eval=TRUE, include=FALSE}
stats = readRDS("FigureS3_statistics.rds")

ft_list = list()
for(gene in names(stats)) {
  for(subtype in names(stats[[gene]])) {
    tmp=capture.output(print(stats[[gene]][[subtype]], print.n="records", show.rmean=FALSE))
    tmp[4] = "Group                                    n events median 0.95LCL 0.95UCL"
    tmp[5:length(tmp)] = tmp[5:length(tmp)] %>% regexPipes::gsub("Group\\=", "") %>% regexPipes::gsub("Low ", "Low") %>% regexPipes::gsub("Neutral ", "Neutral") %>% regexPipes::gsub("High ", "High")
    table = read.table(textConnection(tmp), skip=3, header=TRUE)
    table[,1] = table[,1] %>% regexPipes::gsub("Low", "Low ") %>% regexPipes::gsub("Neutral", "Neutral ") %>% regexPipes::gsub("High", "High ")
    colnames(table) = colnames(table) %>% regexPipes::gsub("X0\\.95", "0\\.95\\% ") %>% regexPipes::gsub("events", "Events") %>% regexPipes::gsub("median", "Median")  
    
    ft = flextable(table)
    ft = add_header_lines(ft, values = subtype)
    ft = add_header_lines(ft, values = gene)
    ft_list[[gene]][[subtype]] = autofit(ft)
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC
```

### Table S6 (Figure S4 - Survival analysis - Expression groups)
```{r echo = FALSE, eval=TRUE, include=FALSE}
stats = readRDS("FigureS4_statistics.rds")

ft_list = list()
for(gene in names(stats)) {
  for(subtype in names(stats[[gene]])) {
    tmp=capture.output(print(stats[[gene]][[subtype]], print.n="records", show.rmean=FALSE))
    tmp[3] = "Group                                    n events median 0.95LCL 0.95UCL"
    tmp[4:length(tmp)] = tmp[4:length(tmp)] %>% regexPipes::gsub("Group\\=", "") %>% regexPipes::gsub("Low ", "Low") %>% regexPipes::gsub("Neutral ", "Neutral") %>% regexPipes::gsub("High ", "High")
    table = read.table(textConnection(tmp), skip=2, header=TRUE)
    table[,1] = table[,1] %>% regexPipes::gsub("Low", "Low ") %>% regexPipes::gsub("Neutral", "Neutral ") %>% regexPipes::gsub("High", "High ")
    colnames(table) = colnames(table) %>% regexPipes::gsub("X0\\.95", "0\\.95\\% ") %>% regexPipes::gsub("events", "Events") %>% regexPipes::gsub("median", "Median")  
    
    ft = flextable(table)
    ft = add_header_lines(ft, values = subtype)
    ft = add_header_lines(ft, values = gene)
    ft_list[[gene]][[subtype]] = autofit(ft)
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC
```

### Table S7 (Figure S9 - Node count between CNA groups) - Inferential statistics
```{r echo = FALSE, eval=TRUE, include=FALSE}
## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")
# Build the table of results from the negative binomial model (for nodes.)
col_names = c("Factor level", "Estimate", "Std. error", "z-value", "p-value")

ft_list = list()
for(sig_gene in genes_sig) {
  results = readRDS(paste0("FigureS9_", sig_gene, "_inferential_statistics.rds"))
  
  ft_list[[sig_gene]] = list()
  for(subtype in names(results$Nodes$Test)) {
    coef_formatted = as.data.frame(apply(coef(results$Nodes$Test[[subtype]]), c(1,2), signif, digits = 2))
    rownames(coef_formatted) = base::gsub("UbSig", "", rownames(coef_formatted))
    coef_formatted = add_column(coef_formatted, d = rownames(coef_formatted), .before = "Estimate")
    colnames(coef_formatted) = col_names
    
    significance = c()
    for(p_value in coef_formatted$`p-value`) {
      if(p_value < 0.001) {
        significance_i = "***"
      } else if(p_value >= 0.001 & p_value < 0.01) {
        significance_i = "**"
      } else if(p_value >= 0.01 & p_value < 0.05) {
        significance_i = "*"
      } else {
        significance_i = ""
      }
      
      significance = c(significance, significance_i)
    }
    
    coef_formatted$Significance = significance  
    ft = flextable(coef_formatted)
    ft = add_header_lines(ft, values = subtype)
    ft = add_header_lines(ft, values = sig_gene)
    ft_list[[sig_gene]][[subtype]] = autofit(ft)
    
  }
}
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC
```

### Table S8 (Figure S10 - Node count between expression groups) - Inferential statistics
```{r echo = FALSE, eval=TRUE, include=FALSE}
## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")
# Build the table of results from the negative binomial model (for nodes.)
col_names = c("Factor level", "Estimate", "Std. error", "z-value", "p-value")

ft_list = list()
for(sig_gene in genes_sig) {
  results = readRDS(paste0("FigureS10_", sig_gene, "_inferential_statistics.rds"))
  
  ft_list[[sig_gene]] = list()
  for(subtype in names(results$Nodes$Test)) {
    coef_formatted = as.data.frame(apply(coef(results$Nodes$Test[[subtype]]), c(1,2), signif, digits = 2))
    rownames(coef_formatted) = base::gsub("UbSig", "", rownames(coef_formatted))
    coef_formatted = add_column(coef_formatted, d = rownames(coef_formatted), .before = "Estimate")
    colnames(coef_formatted) = col_names
    
    significance = c()
    for(p_value in coef_formatted$`p-value`) {
      if(p_value < 0.001) {
        significance_i = "***"
      } else if(p_value >= 0.001 & p_value < 0.01) {
        significance_i = "**"
      } else if(p_value >= 0.01 & p_value < 0.05) {
        significance_i = "*"
      } else {
        significance_i = ""
      }
      
      significance = c(significance, significance_i)
    }
    
    coef_formatted$Significance = significance  
    ft = flextable(coef_formatted)
    ft = add_header_lines(ft, values = subtype)
    ft = add_header_lines(ft, values = sig_gene)
    ft_list[[sig_gene]][[subtype]] = autofit(ft)
    
  }
}
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC

```

### Table XX (Figure S9 - Node counts between CNA groups) - Descriptive statistics
```{r echo = FALSE, eval=TRUE, include=FALSE}
## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")
# Build the table of results from the negative binomial model (for nodes.)
col_names = c("Factor level", "Estimate", "Std. error", "z-value", "p-value")

ft_list = list()
for(sig_gene in genes_sig) {
  results = readRDS(paste0("FigureS9_", sig_gene, "_descriptive_statistics.rds"))
  
  descript_stats = data.frame(
    Subtype = character(),
    MeanPosNodes = numeric(),
    Variance = numeric()
  )  
  
  table = do.call(rbind, results)
  
  ft = flextable(table)
  ft = add_header_lines(ft, values = sig_gene)
  ft_list[[sig_gene]] = autofit(ft)
  
}
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2
ft_list$USP10
ft_list$USP13
ft_list$FZR1
```

### Table XX (Figure S10 - Node counts between expression groups) - Descriptive statistics
```{r echo = FALSE, eval=TRUE, include=FALSE}
## Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")
# Build the table of results from the negative binomial model (for nodes.)
col_names = c("Factor level", "Estimate", "Std. error", "z-value", "p-value")

ft_list = list()
for(sig_gene in genes_sig) {
  results = readRDS(paste0("FigureS10_", sig_gene, "_descriptive_statistics.rds"))
  
  descript_stats = data.frame(
    Subtype = character(),
    MeanPosNodes = numeric(),
    Variance = numeric()
  )  
  
  table = do.call(rbind, results)
  
  ft = flextable(table)
  ft = add_header_lines(ft, values = sig_gene)
  ft_list[[sig_gene]] = autofit(ft)
  
}
```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2
ft_list$USP10
ft_list$USP13
ft_list$FZR1

```

### Table S9 (Figure S11)
```{r echo = FALSE, eval = TRUE, fig.cap=""}
# Set the genes for the signature.
genes_sig = c("SKP2", "USP10", "USP13", "FZR1")

ft_list = list()
for(j in 1:length(genes_sig)) {
  gene = genes_sig[j]
  
  # Load the statistics.
  stats = readRDS(paste0("FigureS11", letters[j], "_statistics.rds"))
  
  for(subtype in names(stats[[gene]])) {
   descriptive_stats = stats[[gene]][[subtype]]$Summary %>% as.data.frame
   ft = flextable(descriptive_stats)
   ft = add_header_lines(ft, values = subtype)
   ft = add_header_lines(ft, values = gene, top = TRUE)
   ft_list[[gene]][[subtype]] = autofit(ft) 
  
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$SKP2$All
ft_list$SKP2$Luminal
ft_list$SKP2$HER2
ft_list$SKP2$TNBC

ft_list$USP10$All
ft_list$USP10$Luminal
ft_list$USP10$HER2
ft_list$USP10$TNBC

ft_list$USP13$All
ft_list$USP13$Luminal
ft_list$USP13$HER2
ft_list$USP13$TNBC

ft_list$FZR1$All
ft_list$FZR1$Luminal
ft_list$FZR1$HER2
ft_list$FZR1$TNBC
```

### Table S10 (Figure S12)
```{r echo = FALSE, eval = TRUE, fig.cap=""}
# Load the statistics.
stats = readRDS("FigureS12_statistics.rds")

ft_list = list()
for(subtype in names(stats)) {
  for(gene_list in names(stats[[subtype]])) {
   descriptive_stats = stats[[subtype]][[gene_list]]$Summary %>% as.data.frame
   ft = flextable(descriptive_stats)
   ft = add_header_lines(ft, values = gene_list)
   ft = add_header_lines(ft, values = subtype, top = TRUE)
   ft_list[[subtype]][[gene_list]] = autofit(ft) 
  }
}

```

```{r echo = FALSE, eval = TRUE, fig.margin = TRUE, fig.cap = ""}
ft_list$All$`USP13, FZR1`
ft_list$All$`USP10, FZR1`
ft_list$All$`USP14, FZR1`
ft_list$Luminal$`USP13, FZR1`
ft_list$Luminal$`USP10, FZR1`
ft_list$Luminal$`USP14, FZR1`
ft_list$HER2$`USP13, FZR1`
ft_list$HER2$`USP10, FZR1`
ft_list$HER2$`USP14, FZR1`
ft_list$TNBC$`USP13, FZR1`
ft_list$TNBC$`USP10, FZR1`
ft_list$TNBC$`USP14, FZR1`
```

# Discussion
```{r echo=FALSE, eval=TRUE, include=FALSE}
# https://rstudio-pubs-static.s3.amazonaws.com/13301_6641d73cfac741a59c0a851feb99e98b.html

# Get the expression data.
genes = c("USP10", "FZR1", "SKP2")
cna = getProfileData(mycgds, genes, "brca_metabric_cna", "brca_metabric_cna")
# Get the clinical data.
clin_sig = getClinicalData(mycgds, "brca_metabric_cna")
clin_sig$SAMPLE_ID = rownames(clin_sig)
subtypes = list()
subtypes[["Luminal"]] = clin_sig %>% filter(ER_STATUS=="Positive" | PR_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["HER2"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Positive") %>% .$SAMPLE_ID
#subtypes[["Non-TNBC"]] = clin_sig %>% filter(ER_STATUS == "Positive" | PR_STATUS == "Positive" | HER2_STATUS == "Positive") %>% .$SAMPLE_ID
subtypes[["TNBC"]] = clin_sig %>% filter(ER_STATUS == "Negative" & PR_STATUS == "Negative" & HER2_STATUS == "Negative") %>% .$SAMPLE_ID
subtypes[["All"]] = clin_sig %>% .$SAMPLE_ID

plots = list()
for(subtype in names(subtypes)) {
  # Get the sample IDs for this subtype.
  sample_IDs = subtypes[[subtype]]
  # Subset cna.
  cna_subtype = cna[rownames(cna) %in% sample_IDs,,drop=F]
  
  ## Divide up the samples by copy number of the genes in the signature. 
  cna_subtype_ub_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample_
  for(k in 1:nrow(cna_subtype)) {
        if(cna_subtype[k,1] < cna_subtype[k,3]) { cna_subtype_ub_group[k] = paste0("Low ubiquitination")} 
        else if(cna_subtype[k,1] > cna_subtype[k,3]) { cna_subtype_ub_group[k] = paste0("High ubiquitination")}
        else if(cna_subtype[k,1] == cna_subtype[k,3]) { cna_subtype_ub_group[k] = cna_subtype_ub_group[k] = paste0("Intermediate ubiquitination")}
        else {cna_subtype_ub_group[k] = NA}
      }
  cna_subtype$UbGroup = cna_subtype_ub_group

  ## Divide up the samples by SKP2 copy number. 
  cna_subtype_skp2_group = c() # The vector that holds the grouping (high vs_ low expression) for each sample_
  for(k in 1:nrow(cna_subtype)) {
        if(cna_subtype[k,2] < 0) { cna_subtype_skp2_group[k] = paste0("Low SKP2")} 
        else if(cna_subtype[k,2] > 0) { cna_subtype_skp2_group[k] = paste0("High SKP2")}
        else if(cna_subtype[k,2] == 0) { cna_subtype_skp2_group[k] = paste0("Intermediate SKP2")}
        else {cna_subtype_skp2_group[k] = NA}
      }
  cna_subtype$SKP2Group = cna_subtype_skp2_group

  # Get counts of each group.
  high_SKP2_count = sum(cna_subtype$SKP2Group=="High SKP2")
  low_SKP2_count = sum(cna_subtype$SKP2Group=="Low SKP2")
  high_ub_count = sum(cna_subtype$UbGroup=="High ubiquitination")
  low_ub_count = sum(cna_subtype$UbGroup=="Low ubiquitination")
  high_SKP2_low_ub_overlap = nrow(cna_subtype[cna_subtype$SKP2Group=="High SKP2" & cna_subtype$UbGroup=="Low ubiquitination",])
  low_SKP2_high_ub_overlap = nrow(cna_subtype[cna_subtype$SKP2Group=="Low SKP2" & cna_subtype$UbGroup=="High ubiquitination",])
  
  # Create the euler objects.
  e1 = euler(c(
    "A" = high_SKP2_count,
    "B" = low_ub_count,
    "A&B" = high_SKP2_low_ub_overlap
  ))
  e2 = euler(c(
    "A" = low_SKP2_count,
    "B" = high_ub_count,
    "A&B" = low_SKP2_high_ub_overlap
  ))
  plot_e1 = plot(e1, 
                 #main = as.expression(bquote(bold(.(subtype)))),
                 labels = list(labels = c("High SKP2", "Low Ubiquitination"), fontsize = plot_font_size_big), 
                 quantities = list(type=c("counts", "percent"), fontsize = plot_font_size_big), 
                 fills = pal_jco()(4)[c(4,2)]
                 )
  
  if(subtype!="Luminal") {
    plot_e2 = plot(e2, 
                 #main = as.expression(bquote(bold(.(subtype)))),
                 labels = list(labels = c("Low SKP2", "High Ubiquitination"), fontsize = 20), 
                 quantities = list(type=c("counts", "percent"), fontsize = c(plot_font_size_big, plot_font_size_big, plot_font_size_small)), 
                 fills = pal_jco()(7)[c(5,1)]
                 )
  } else {
    plot_e2 = plot(e2, 
                 #main = as.expression(bquote(bold(.(subtype)))),
                 labels = list(labels = c("Low SKP2", "High Ubiquitination"), fontsize = c(plot_font_size_small, plot_font_size_big), y = c(2, 0, 0)), 
                 quantities = list(type=c("counts", "percent"), fontsize = c(plot_font_size_small, plot_font_size_big, 8)), 
                 fills = pal_jco()(7)[c(5,1)]
                 )
  }
  
  
  plots[["e1"]][[subtype]] = plot_e1
  plots[["e2"]][[subtype]] = plot_e2
  
}

# Save to EPS.
setEPS()
postscript("Figure_Discussion.eps", width = 20, height = 20)
ggpubr::ggarrange(plotlist = plots[["e1"]], 
                   nrow = 2, 
                   ncol = 2, 
                   labels = c("Luminal", "HER2", "TNBC", "All"),
                  font.label = list(size = 20)
                   #hjust = c(-0.5, -1, -0.5, -1, -1.5)
                  )
dev.off()
```

```{r echo=FALSE, eval=TRUE, fig.width=15, fig.height=15, fig.margin = TRUE, fig.cap = ""}
# High SKP2, low ubiquitination overlap.
ggpubr::ggarrange(plotlist = plots[["e1"]], 
                   nrow = 2, 
                   ncol = 2, 
                   labels = c("Luminal", "HER2", "TNBC", "All"),
                  font.label = list(size = 20)
                   #hjust = c(-0.5, -1, -0.5, -1, -1.5)
                  )
#invisible(draw.pairwise.venn(high_SKP2_count, low_ub_count, high_SKP2_low_ub_overlap, 
#                   category = c("High SKP2", "Low SKP2 ubiquitination"), 
#                   lty = rep("blank", 2), 
#                   fill = pal_jco()(4)[c(4,2)], 
#                   alpha = c(1, 0.5), #rep(0.5, 2), 
#                   cat.pos = c(0, 0), 
#                   cat.dist = rep(0.025, 2),
#                   fontfamily = "Arial",
#                   cat.fontfamily = "Arial"))

```

```{r echo=FALSE, eval=TRUE, fig.width=15, fig.height=15, fig.margin = TRUE, fig.cap = ""}
# Low SKP2, high ubiquitination overlap.
ggpubr::ggarrange(plotlist = plots[["e2"]], 
                   nrow = 2, 
                   ncol = 2, 
                   labels = c("Luminal", "HER2", "TNBC", "All"),
                  font.label = list(size = 20)
                   #hjust = c(-0.5, -1, -0.5, -1, -1.5)
                  )

#invisible(draw.pairwise.venn(low_SKP2_count, high_ub_count, low_SKP2_high_ub_overlap, 
#                   category = c("Low SKP2", "High SKP2 ubiquitination"), 
#                   lty = rep("blank", 2), 
#                   fill = pal_jco()(7)[c(7,1)], 
#                   alpha = c(0.5, 1), 
#                   cat.pos = c(0, 180), 
#                   cat.dist = rep(0.025, 2),
#                   fontfamily = "Arial",
#                   cat.fontfamily = "Arial"))
```